<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.5.53">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="John Zobolas, Lukas Burk">
<meta name="dcterms.date" content="2024-07-08">
<meta name="description" content="Tutorial for the useR! 2024 conference in Salzburg, Austria (8-11 July)">

<title>Intro to Machine Learning for Survival Analysis with mlr3</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
  margin-bottom: 0em;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>

  <script>window.backupDefine = window.define; window.define = undefined;</script><script src="https://cdn.jsdelivr.net/npm/katex@latest/dist/katex.min.js"></script>
  <script>document.addEventListener("DOMContentLoaded", function () {
 var mathElements = document.getElementsByClassName("math");
 var macros = [];
 for (var i = 0; i < mathElements.length; i++) {
  var texText = mathElements[i].firstChild;
  if (mathElements[i].tagName == "SPAN") {
   katex.render(texText.data, mathElements[i], {
    displayMode: mathElements[i].classList.contains('display'),
    throwOnError: false,
    macros: macros,
    fleqn: false
   });
}}});
  </script>
  <script>window.define = window.backupDefine; window.backupDefine = undefined;</script><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/katex@latest/dist/katex.min.css">

<script type="text/javascript">
const typesetMath = (el) => {
  if (window.MathJax) {
    // MathJax Typeset
    window.MathJax.typeset([el]);
  } else if (window.katex) {
    // KaTeX Render
    var mathElements = el.getElementsByClassName("math");
    var macros = [];
    for (var i = 0; i < mathElements.length; i++) {
      var texText = mathElements[i].firstChild;
      if (mathElements[i].tagName == "SPAN") {
        window.katex.render(texText.data, mathElements[i], {
          displayMode: mathElements[i].classList.contains('display'),
          throwOnError: false,
          macros: macros,
          fleqn: false
        });
      }
    }
  }
}
window.Quarto = {
  typesetMath
};
</script>

</head>

<body class="floating nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" role="menu" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link active" href="./index.html" aria-current="page"> 
<span class="menu-text">ML Survival with mlr3</span></a>
  </li>  
</ul>
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item compact">
    <a class="nav-link" href="https://github.com/survival-org/user2024-survival-mlr3"> <i class="bi bi-github" role="img">
</i> 
<span class="menu-text"></span></a>
  </li>  
</ul>
          </div> <!-- /navcollapse -->
            <div class="quarto-navbar-tools">
</div>
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-full page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto">
    <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#mlr3-basics" id="toc-mlr3-basics" class="nav-link active" data-scroll-target="#mlr3-basics"><code>mlr3</code>: Basics</a></li>
  <li><a href="#example-train-predict-evaluate" id="toc-example-train-predict-evaluate" class="nav-link" data-scroll-target="#example-train-predict-evaluate">Example: Train-Predict-Evaluate</a></li>
  <li><a href="#mlr3proba-basics" id="toc-mlr3proba-basics" class="nav-link" data-scroll-target="#mlr3proba-basics"><code>mlr3proba</code>: Basics</a>
  <ul class="collapse">
  <li><a href="#survival-tasks" id="toc-survival-tasks" class="nav-link" data-scroll-target="#survival-tasks">Survival Tasks</a></li>
  <li><a href="#coxph-learner" id="toc-coxph-learner" class="nav-link" data-scroll-target="#coxph-learner">CoxPH learner</a></li>
  <li><a href="#prediction-types" id="toc-prediction-types" class="nav-link" data-scroll-target="#prediction-types">Prediction types</a></li>
  <li><a href="#model-evaluation" id="toc-model-evaluation" class="nav-link" data-scroll-target="#model-evaluation">Model evaluation</a></li>
  </ul></li>
  <li><a href="#using-and-tuning-ml-survival-models-on-high-dimensional-data" id="toc-using-and-tuning-ml-survival-models-on-high-dimensional-data" class="nav-link" data-scroll-target="#using-and-tuning-ml-survival-models-on-high-dimensional-data">Using and tuning ML survival models on high-dimensional data</a></li>
  <li><a href="#benchmarking-with-multiple-datasets" id="toc-benchmarking-with-multiple-datasets" class="nav-link" data-scroll-target="#benchmarking-with-multiple-datasets">Benchmarking with multiple datasets</a>
  <ul class="collapse">
  <li><a href="#statistical-analysis" id="toc-statistical-analysis" class="nav-link" data-scroll-target="#statistical-analysis">Statistical analysis</a></li>
  </ul></li>
  <li><a href="#conclusion" id="toc-conclusion" class="nav-link" data-scroll-target="#conclusion">Conclusion</a></li>
  <li><a href="#session-info" id="toc-session-info" class="nav-link" data-scroll-target="#session-info">Session Info</a></li>
  </ul>
<div class="toc-actions"><ul><li><a href="https://github.com/survival-org/user2024-survival-mlr3/blob/main/index.qmd" class="toc-action"><i class="bi bi-github"></i>View source</a></li></ul></div></nav>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar zindex-bottom">
    </div>
<!-- main -->
<main class="content column-page-right" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Intro to Machine Learning for Survival Analysis with mlr3</h1>
</div>

<div>
  <div class="description">
    Tutorial for the useR! 2024 conference in Salzburg, Austria (8-11 July)
  </div>
</div>


<div class="quarto-title-meta column-page-right">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p><a href="https://github.com/bblodfon">John Zobolas</a>, <a href="https://lukasburk.de/">Lukas Burk</a> </p>
          </div>
  </div>
    
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">July 8, 2024</p>
    </div>
  </div>
  
    
  </div>
  


</header>


<section id="mlr3-basics" class="level2 unnumbered">
<h2 class="unnumbered anchored" data-anchor-id="mlr3-basics"><code>mlr3</code>: Basics</h2>
<div class="callout callout-style-default callout-tip callout-titled" title="Teaching Aims">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Teaching Aims
</div>
</div>
<div class="callout-body-container callout-body">
<ul>
<li>Understand how <code>{mlr3}</code> is structured <span class="citation" data-cites="mlr3">(<a href="#ref-mlr3" role="doc-biblioref">Lang et al. 2019</a>)</span></li>
<li>Access <code>learner</code>s and (built-in) <code>task</code>s</li>
</ul>
</div>
</div>
<p>To get started, we load <code>{mlr3verse}</code>, which will load various packages from the <code>{mlr3}</code> ecosystem:</p>
<div class="cell">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(mlr3verse)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p><code>{mlr3}</code> ships with wrappers for many commonly used machine learning algorithms (“learners”).<br>
We can access the list of available learners using the <code>mlr_learners</code> dictionary:</p>
<div class="cell">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="fu">sample</span>(mlr_learners<span class="sc">$</span><span class="fu">keys</span>(), <span class="dv">10</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "classif.decision_table" "classif.abess"          "surv.bart"             
 [4] "clust.diana"            "classif.ctree"          "classif.catboost"      
 [7] "classif.random_tree"    "dens.nonpar"            "classif.earth"         
[10] "surv.mboost"           </code></pre>
</div>
</div>
<p>One example:</p>
<div class="cell">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="fu">lrn</span>(<span class="st">"classif.ranger"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>&lt;LearnerClassifRanger:classif.ranger&gt;: Random Forest
* Model: -
* Parameters: num.threads=1
* Packages: mlr3, mlr3learners, ranger
* Predict Types:  [response], prob
* Feature Types: logical, integer, numeric, character, factor, ordered
* Properties: hotstart_backward, importance, multiclass, oob_error,
  twoclass, weights</code></pre>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>Use <code>lrn("classif.ranger")$help()</code> to view the help page, with links to documentation for parameters and other information about the wrapped learner.</p>
</div>
</div>
<p>Built-in tasks can be accessed using the <code>mlr_tasks</code> dictionary:</p>
<div class="cell">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(<span class="fu">as.data.table</span>(mlr_tasks)[, <span class="fu">list</span>(key, label, task_type, nrow, ncol, properties)])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>Key: &lt;key&gt;
              key                     label task_type  nrow  ncol properties
           &lt;char&gt;                    &lt;char&gt;    &lt;char&gt; &lt;int&gt; &lt;int&gt;     &lt;list&gt;
1:   ames_housing          Ames House Sales      regr  2930    82           
2:   bike_sharing       Bike Sharing Demand      regr 17379    14           
3: boston_housing     Boston Housing Prices      regr   506    18           
4:  breast_cancer   Wisconsin Breast Cancer   classif   683    10   twoclass
5:  german_credit             German Credit   classif  1000    21   twoclass
6:           ilpd Indian Liver Patient Data   classif   583    11   twoclass</code></pre>
</div>
</div>
<p>One example:</p>
<div class="cell">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="fu">tsk</span>(<span class="st">"penguins_simple"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>&lt;TaskClassif:penguins&gt; (333 x 11): Simplified Palmer Penguins
* Target: species
* Properties: multiclass
* Features (10):
  - dbl (7): bill_depth, bill_length, island.Biscoe, island.Dream,
    island.Torgersen, sex.female, sex.male
  - int (3): body_mass, flipper_length, year</code></pre>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>Tasks encapsulate a data source (typically a <code>data.table</code>) and additional information regarding which variables are considered features and target. Tasks can also specify additional properties such as stratification, which we will see later.</p>
</div>
</div>
</section>
<section id="example-train-predict-evaluate" class="level2 unnumbered">
<h2 class="unnumbered anchored" data-anchor-id="example-train-predict-evaluate">Example: Train-Predict-Evaluate</h2>
<div class="callout callout-style-default callout-tip callout-titled" title="Teaching Aims">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Teaching Aims
</div>
</div>
<div class="callout-body-container callout-body">
<ul>
<li>Perform a simple train-predict-evaluate step</li>
<li>Use built-in classification <code>task</code> and <code>learner</code></li>
</ul>
</div>
</div>
<p>The below code snippet trains a random forest model on the <code>penguins_simple</code> task (a simplified version of the <code>palmerpenguins</code> dataset, but without missing values) and evaluates the model’s performance using the classification error metric:</p>
<div class="cell">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>task <span class="ot">=</span> <span class="fu">tsk</span>(<span class="st">"penguins_simple"</span>)</span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>learner <span class="ot">=</span> <span class="fu">lrn</span>(<span class="st">"classif.ranger"</span>, <span class="at">num.trees =</span> <span class="dv">10</span>)</span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a>part <span class="ot">=</span> <span class="fu">partition</span>(task, <span class="at">ratio =</span> <span class="fl">0.8</span>) <span class="co"># by default stratifies on the target column</span></span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a>learner<span class="sc">$</span><span class="fu">train</span>(task, <span class="at">row_ids =</span> part<span class="sc">$</span>train)</span>
<span id="cb10-7"><a href="#cb10-7" aria-hidden="true" tabindex="-1"></a>preds <span class="ot">=</span> learner<span class="sc">$</span><span class="fu">predict</span>(task, <span class="at">row_ids =</span> part<span class="sc">$</span>test)</span>
<span id="cb10-8"><a href="#cb10-8" aria-hidden="true" tabindex="-1"></a>preds<span class="sc">$</span><span class="fu">score</span>(<span class="fu">msr</span>(<span class="st">"classif.ce"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>classif.ce 
0.01492537 </code></pre>
</div>
</div>
<p>Learn more by reading the <a href="https://mlr3book.mlr-org.com/data_and_basic_modeling.html">respective chapter</a> on the <code>mlr3</code> book.</p>
</section>
<section id="mlr3proba-basics" class="level2 unnumbered">
<h2 class="unnumbered anchored" data-anchor-id="mlr3proba-basics"><code>mlr3proba</code>: Basics</h2>
<div class="callout callout-style-default callout-tip callout-titled" title="Teaching Aims">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Teaching Aims
</div>
</div>
<div class="callout-body-container callout-body">
<ul>
<li>Understand survival tasks and how they differ from regression/classification</li>
<li>Know how to conduct basic modeling with <code>{mlr3proba}</code> <span class="citation" data-cites="mlr3proba">(<a href="#ref-mlr3proba" role="doc-biblioref">Sonabend et al. 2021</a>)</span></li>
<li>Prediction types</li>
<li>Survival measures</li>
</ul>
</div>
</div>
<p><code>{mlr3proba}</code> extends <code>{mlr3}</code> with survival analysis capabilities.</p>
<div class="callout callout-style-default callout-important callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Important
</div>
</div>
<div class="callout-body-container callout-body">
<p>As of now, <code>{mlr3proba}</code> is not on CRAN, but you can install it <a href="https://github.com/mlr-org/mlr3proba/?tab=readme-ov-file#installation">from GitHub</a> or <a href="https://mlr-org.r-universe.dev/mlr3proba">r-universe</a>. More info is also available on the respective <a href="https://mlr3book.mlr-org.com/chapters/chapter13/beyond_regression_and_classification.html#sec-survival">mlr3 book chapter</a>.</p>
</div>
</div>
<section id="survival-tasks" class="level3 unnumbered">
<h3 class="unnumbered anchored" data-anchor-id="survival-tasks">Survival Tasks</h3>
<p>We’ll start by using the built-in <code>lung</code> dataset, which is a survival task with <span class="math inline">7</span> features and <span class="math inline">168</span> observations:</p>
<div class="cell">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(mlr3proba)</span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a>task <span class="ot">=</span> <span class="fu">tsk</span>(<span class="st">"lung"</span>)</span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a>task</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>&lt;TaskSurv:lung&gt; (168 x 9): Lung Cancer
* Target: time, status
* Properties: -
* Features (7):
  - int (6): age, meal.cal, pat.karno, ph.ecog, ph.karno, wt.loss
  - fct (1): sex</code></pre>
</div>
</div>
<p><a href="https://mlr3proba.mlr-org.com/reference/TaskSurv.html#methods">See online reference</a> to useful methods offered by the main <code>TaskSurv</code> class. Some examples:</p>
<p>Target <code>Surv</code> object from <code>{survival}</code> (<code>+</code> denotes censored observation):</p>
<div class="cell">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(task<span class="sc">$</span><span class="fu">truth</span>())</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>[1]  455   210  1022+  310   361   218 </code></pre>
</div>
</div>
<p>Proportion of censored observations:</p>
<div class="cell">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a>task<span class="sc">$</span><span class="fu">cens_prop</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0.2797619</code></pre>
</div>
</div>
<p>Does the data satisfy the <strong>proportional hazards</strong> assumption? Get the p-value from the Grambsch-Therneau test (see <code>?survival::cox.zph</code> <span class="citation" data-cites="Grambsch1994">(<a href="#ref-Grambsch1994" role="doc-biblioref">Grambsch and Therneau 1994</a>)</span>):</p>
<div class="cell">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a>task<span class="sc">$</span><span class="fu">prop_haz</span>() <span class="co"># barely, p &gt; 0.05 =&gt; PH</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0.0608371</code></pre>
</div>
</div>
<p>Using the <code>autoplot()</code> function from <code>{ggplot2}</code>, we get the Kaplan-Meier curve:</p>
<div class="cell">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a><span class="fu">autoplot</span>(task) <span class="sc">+</span></span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"Lung dataset: Kaplan-Meier curve"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stderr">
<pre><code>Registered S3 method overwritten by 'GGally':
  method from   
  +.gg   ggplot2</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="index_files/figure-html/km-curve-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Tasks shipped with <code>{mlr3proba}</code>:</p>
<div class="cell">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="fu">as.data.table</span>(mlr_tasks)[task_type <span class="sc">==</span> <span class="st">"surv"</span>, <span class="fu">list</span>(key, label, nrow, ncol)]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>Key: &lt;key&gt;
             key                       label  nrow  ncol
          &lt;char&gt;                      &lt;char&gt; &lt;int&gt; &lt;int&gt;
 1:         actg                    ACTG 320  1151    13
 2:         gbcs        German Breast Cancer   686    10
 3:         gbsg        German Breast Cancer   686    10
 4:        grace                  GRACE 1000  1000     8
 5:         lung                 Lung Cancer   168     9
 6:         mgus                        MGUS   176     9
 7:          pbc Primary Biliary Cholangitis   276    19
 8:         rats                        Rats   300     5
 9: unemployment       Unemployment Duration  3343     6
10:      veteran                     Veteran   137     8
11:         whas      Worcester Heart Attack   481    11</code></pre>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<ul>
<li>Use <a href="https://mlr3proba.mlr-org.com/reference/as_task_surv.html">as_task_surv()</a> to convert your own datasets to a <code>TaskSurv</code> object</li>
<li>Try <code>tsk("lung")$help()</code> to get more info about the dataset and pre-processing applied</li>
</ul>
</div>
</div>
</section>
<section id="coxph-learner" class="level3 unnumbered">
<h3 class="unnumbered anchored" data-anchor-id="coxph-learner">CoxPH learner</h3>
<p>The classical Cox Proportional Hazards model:</p>
<div class="cell">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a>cox <span class="ot">=</span> <span class="fu">lrn</span>(<span class="st">"surv.coxph"</span>)</span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a>cox</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>&lt;LearnerSurvCoxPH:surv.coxph&gt;: Cox Proportional Hazards
* Model: -
* Parameters: list()
* Packages: mlr3, mlr3proba, survival, distr6
* Predict Types:  [crank], distr, lp
* Feature Types: logical, integer, numeric, factor
* Properties: weights</code></pre>
</div>
</div>
<p>Train the cox model and access the fit object from the <code>{survival}</code> package:</p>
<div class="cell">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">42</span>)</span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a>part <span class="ot">=</span> <span class="fu">partition</span>(task, <span class="at">ratio =</span> <span class="fl">0.8</span>) <span class="co"># by default, stratification is on `status` variable</span></span>
<span id="cb26-3"><a href="#cb26-3" aria-hidden="true" tabindex="-1"></a>cox<span class="sc">$</span><span class="fu">train</span>(task, <span class="at">row_ids =</span> part<span class="sc">$</span>train)</span>
<span id="cb26-4"><a href="#cb26-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-5"><a href="#cb26-5" aria-hidden="true" tabindex="-1"></a>cox<span class="sc">$</span>model</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>Call:
survival::coxph(formula = task$formula(), data = task$data(), 
    x = TRUE)

                coef  exp(coef)   se(coef)      z      p
age        1.341e-02  1.013e+00  1.258e-02  1.066 0.2864
meal.cal  -5.007e-05  9.999e-01  2.903e-04 -0.172 0.8631
pat.karno -2.142e-02  9.788e-01  9.055e-03 -2.366 0.0180
ph.ecog    5.936e-01  1.811e+00  2.500e-01  2.375 0.0176
ph.karno   2.541e-02  1.026e+00  1.263e-02  2.011 0.0443
sexm       4.510e-01  1.570e+00  2.298e-01  1.962 0.0497
wt.loss   -1.500e-02  9.851e-01  8.395e-03 -1.787 0.0739

Likelihood ratio test=23.36  on 7 df, p=0.001475
n= 135, number of events= 97 </code></pre>
</div>
</div>
<p>Visual output of the model, using the latest version from Github of <code>{mlr3viz}</code>:</p>
<div class="cell">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a><span class="fu">autoplot</span>(cox)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="index_files/figure-html/cox-model-viz-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="prediction-types" class="level3 unnumbered">
<h3 class="unnumbered anchored" data-anchor-id="prediction-types">Prediction types</h3>
<p>Let’s predict using the trained cox model on the test set (output is a <a href="https://mlr3proba.mlr-org.com/reference/PredictionSurv.html">PredictionSurv</a> object):</p>
<div class="cell">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a>p <span class="ot">=</span> cox<span class="sc">$</span><span class="fu">predict</span>(task, <span class="at">row_ids =</span> part<span class="sc">$</span>test)</span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a>p</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>&lt;PredictionSurv&gt; for 33 observations:
    row_ids time status       crank          lp     distr
          1  455   TRUE -0.16022736 -0.16022736 &lt;list[1]&gt;
          8  170   TRUE  0.07608537  0.07608537 &lt;list[1]&gt;
         15  371   TRUE -0.46601841 -0.46601841 &lt;list[1]&gt;
---                                                      
        165  191  FALSE -0.30526841 -0.30526841 &lt;list[1]&gt;
        166  105  FALSE  0.49632782  0.49632782 &lt;list[1]&gt;
        168  177  FALSE -0.17234336 -0.17234336 &lt;list[1]&gt;</code></pre>
</div>
</div>
<div class="callout callout-style-default callout-tip callout-titled" title="Prediction types in mlr3proba">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Prediction types in mlr3proba
</div>
</div>
<div class="callout-body-container callout-body">
<ul>
<li><code>crank</code>: Continuous risk ranking</li>
<li><code>lp</code>: Linear predictor calculated as <span class="math inline">\hat\beta * X_{test}</span></li>
<li><code>distr</code>: Predicted survival distribution, either discrete or continuous</li>
<li><code>response</code>: Predicted survival time</li>
</ul>
</div>
</div>
<p>For the cox model, <code>crank = lp</code> (the higher, the more risk):</p>
<div class="cell">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a>p<span class="sc">$</span>lp</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>           1            2            3            4            5            6 
-0.160227364  0.076085366 -0.466018411  0.293380270  1.179147761  0.523244848 
           7            8            9           10           11           12 
 0.391564618 -0.029833700 -0.149489235 -0.262762070  0.076021387  0.279388934 
          13           14           15           16           17           18 
 0.889995280  0.859467193  1.030472975  0.277533930 -0.057165655  0.362416853 
          19           20           21           22           23           24 
-0.037670338 -0.295071061 -0.419840184  0.793214751  0.823500785  0.977222024 
          25           26           27           28           29           30 
-0.046252611  0.021227170 -0.093541236 -0.158438686  1.615114453  0.003701068 
          31           32           33 
-0.305268413  0.496327822 -0.172343361 </code></pre>
</div>
</div>
<p>Survival prediction is a 2D <code>matrix</code> essentially, with dimensions: <em>observations</em> x <em>time points</em>:</p>
<div class="cell">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a>p<span class="sc">$</span>data<span class="sc">$</span>distr[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>          5        11        12        13        15
1 0.9959775 0.9919519 0.9879072 0.9837970 0.9796477
2 0.9949079 0.9898175 0.9847084 0.9795222 0.9742926
3 0.9970357 0.9940659 0.9910789 0.9880402 0.9849691
4 0.9936759 0.9873617 0.9810323 0.9746156 0.9681535
5 0.9847342 0.9696296 0.9546262 0.9395560 0.9245212</code></pre>
</div>
</div>
<p>Users should use the <a href="https://github.com/xoopR/distr6">distr6</a> interface <span class="citation" data-cites="distr6">(<a href="#ref-distr6" role="doc-biblioref">Sonabend and Kirdly 2021</a>)</span> to access this prediction type, which allows us to retrieve survival probabilities (or hazards) for any time point of interest:</p>
<div class="cell">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a><span class="co"># first 4 patients in the test set, specific time points:</span></span>
<span id="cb35-2"><a href="#cb35-2" aria-hidden="true" tabindex="-1"></a>p<span class="sc">$</span>distr[<span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>]<span class="sc">$</span><span class="fu">survival</span>(<span class="fu">c</span>(<span class="dv">100</span>, <span class="dv">500</span>, <span class="dv">1200</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>          [,1]      [,2]      [,3]       [,4]
100  0.9184997 0.8979186 0.9393041 0.87475634
500  0.4589611 0.3729197 0.5634874 0.29352281
1200 0.1684876 0.1048078 0.2693617 0.06062239</code></pre>
</div>
</div>
<p>Visualization of predicted survival curves for <span class="math inline">3</span> test patients:</p>
<div class="cell">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a>p2 <span class="ot">=</span> p<span class="sc">$</span><span class="fu">clone</span>()<span class="sc">$</span><span class="fu">filter</span>(<span class="at">row_ids =</span> <span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">24</span>,<span class="dv">40</span>))</span>
<span id="cb37-2"><a href="#cb37-2" aria-hidden="true" tabindex="-1"></a><span class="fu">autoplot</span>(p2, <span class="at">type =</span> <span class="st">"preds"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="index_files/figure-html/pred-curves-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="model-evaluation" class="level3 unnumbered">
<h3 class="unnumbered anchored" data-anchor-id="model-evaluation">Model evaluation</h3>
<div class="callout callout-style-default callout-tip callout-titled" title="Model validation">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Model validation
</div>
</div>
<div class="callout-body-container callout-body">
<p>Validation of a survival model can be done by assessing:</p>
<ol type="1">
<li><strong>Discrimination</strong>: the ability of the model to distinguish between low and high risk patients</li>
<li><strong>Calibration</strong>: the agreement between the observed and predicted survival probabilities</li>
<li><strong>Overall performance</strong>: the distance between the observed and predicted survival probabilities</li>
</ol>
</div>
</div>
<p>Many measures included in <code>mlr3proba</code>:</p>
<div class="cell">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb38"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a>mlr_measures<span class="sc">$</span><span class="fu">keys</span>(<span class="at">pattern =</span> <span class="st">"surv"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "surv.brier"         "surv.calib_alpha"   "surv.calib_beta"   
 [4] "surv.chambless_auc" "surv.cindex"        "surv.dcalib"       
 [7] "surv.graf"          "surv.hung_auc"      "surv.intlogloss"   
[10] "surv.logloss"       "surv.mae"           "surv.mse"          
[13] "surv.nagelk_r2"     "surv.oquigley_r2"   "surv.rcll"         
[16] "surv.rmse"          "surv.schmid"        "surv.song_auc"     
[19] "surv.song_tnr"      "surv.song_tpr"      "surv.uno_auc"      
[22] "surv.uno_tnr"       "surv.uno_tpr"       "surv.xu_r2"        </code></pre>
</div>
</div>
<p>Most commonly used metrics are for assessing discrimination, such as <strong>Harrell’s C-index</strong> <span class="citation" data-cites="Harrell1982">(<a href="#ref-Harrell1982" role="doc-biblioref">Harrell et al. 1982</a>)</span>, <strong>Uno’s C-index</strong> <span class="citation" data-cites="Uno2011">(<a href="#ref-Uno2011" role="doc-biblioref">Uno et al. 2011</a>)</span> and the <strong>(time-dependent) AUC</strong> <span class="citation" data-cites="Heagerty2005 Uno2007">(<a href="#ref-Heagerty2005" role="doc-biblioref">Heagerty and Zheng 2005</a>; <a href="#ref-Uno2007" role="doc-biblioref">Uno et al. 2007</a>)</span>:</p>
<div class="cell">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb40"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a>harrell_c <span class="ot">=</span> <span class="fu">msr</span>(<span class="st">"surv.cindex"</span>, <span class="at">id =</span> <span class="st">"surv.cindex.harrell"</span>)</span>
<span id="cb40-2"><a href="#cb40-2" aria-hidden="true" tabindex="-1"></a>uno_c <span class="ot">=</span> <span class="fu">msr</span>(<span class="st">"surv.cindex"</span>, <span class="at">weight_meth =</span> <span class="st">"G2"</span>, <span class="at">id =</span> <span class="st">"surv.cindex.uno"</span>)</span>
<span id="cb40-3"><a href="#cb40-3" aria-hidden="true" tabindex="-1"></a>uno_auci <span class="ot">=</span> <span class="fu">msr</span>(<span class="st">"surv.uno_auc"</span>, <span class="at">integrated =</span> <span class="cn">TRUE</span>) <span class="co"># across all times in the test set</span></span>
<span id="cb40-4"><a href="#cb40-4" aria-hidden="true" tabindex="-1"></a>uno_auc <span class="ot">=</span> <span class="fu">msr</span>(<span class="st">"surv.uno_auc"</span>, <span class="at">integrated =</span> <span class="cn">FALSE</span>, <span class="at">times =</span> <span class="dv">10</span>) <span class="co"># at a specific time-point of interest</span></span>
<span id="cb40-5"><a href="#cb40-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb40-6"><a href="#cb40-6" aria-hidden="true" tabindex="-1"></a>harrell_c</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>&lt;MeasureSurvCindex:surv.cindex.harrell&gt;
* Packages: mlr3, mlr3proba
* Range: [0, 1]
* Minimize: FALSE
* Average: macro
* Parameters: weight_meth=I, tiex=0.5, eps=0.001
* Properties: -
* Predict type: crank
* Return type: Score</code></pre>
</div>
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb42"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a>uno_auc</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>&lt;MeasureSurvUnoAUC:surv.uno_auc&gt;
* Packages: mlr3, mlr3proba, survAUC
* Range: [0, 1]
* Minimize: FALSE
* Average: macro
* Parameters: integrated=FALSE, times=10
* Properties: requires_task, requires_train_set
* Predict type: lp
* Return type: Score</code></pre>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<ul>
<li>Not all measures are applicable to all models - <strong>prediction type</strong> matters!</li>
<li>Most discrimination metrics use the <code>crank</code> or <code>lp</code> prediction</li>
</ul>
</div>
</div>
<div class="cell">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb44"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a>p<span class="sc">$</span><span class="fu">score</span>(harrell_c)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>surv.cindex.harrell 
          0.6336898 </code></pre>
</div>
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb46"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb46-1"><a href="#cb46-1" aria-hidden="true" tabindex="-1"></a>p<span class="sc">$</span><span class="fu">score</span>(uno_c, <span class="at">task =</span> task, <span class="at">train_set =</span> part<span class="sc">$</span>train)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>surv.cindex.uno 
      0.5907828 </code></pre>
</div>
</div>
<p>Calibration is traditionally performed graphically via calibration plots:</p>
<div class="cell">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb48"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb48-1"><a href="#cb48-1" aria-hidden="true" tabindex="-1"></a><span class="fu">autoplot</span>(p, <span class="at">type =</span> <span class="st">"calib"</span>, <span class="at">task =</span> task, <span class="at">row_ids =</span> part<span class="sc">$</span>test)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="index_files/figure-html/calib-plot-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>But there exists also calibration metrics, e.g.&nbsp;<strong>D-Calibration</strong> <span class="citation" data-cites="Haider2020">(<a href="#ref-Haider2020" role="doc-biblioref">Haider et al. 2020</a>)</span>:</p>
<div class="cell">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb49"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb49-1"><a href="#cb49-1" aria-hidden="true" tabindex="-1"></a>dcal <span class="ot">=</span> <span class="fu">msr</span>(<span class="st">"surv.dcalib"</span>)</span>
<span id="cb49-2"><a href="#cb49-2" aria-hidden="true" tabindex="-1"></a>dcal</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>&lt;MeasureSurvDCalibration:surv.dcalib&gt;
* Packages: mlr3, mlr3proba
* Range: [0, Inf]
* Minimize: TRUE
* Average: macro
* Parameters: B=10, chisq=FALSE, truncate=Inf
* Properties: -
* Predict type: distr
* Return type: Score</code></pre>
</div>
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb51"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb51-1"><a href="#cb51-1" aria-hidden="true" tabindex="-1"></a>p<span class="sc">$</span><span class="fu">score</span>(dcal)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>surv.dcalib 
   8.320423 </code></pre>
</div>
</div>
<p>Overall survival prediction performance can be assessed by scoring rules such as the <strong>Integrated Survival Brier Score</strong> (ISBS) <span class="citation" data-cites="Graf1999">(<a href="#ref-Graf1999" role="doc-biblioref">Graf et al. 1999</a>)</span> and the <strong>Right-censored Log-Loss</strong> (RCLL) <span class="citation" data-cites="Avati2020">(<a href="#ref-Avati2020" role="doc-biblioref">Avati et al. 2020</a>)</span> among others:</p>
<div class="cell">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb53"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb53-1"><a href="#cb53-1" aria-hidden="true" tabindex="-1"></a>rcll <span class="ot">=</span> <span class="fu">msr</span>(<span class="st">"surv.rcll"</span>)</span>
<span id="cb53-2"><a href="#cb53-2" aria-hidden="true" tabindex="-1"></a>rcll</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>&lt;MeasureSurvRCLL:surv.rcll&gt;
* Packages: mlr3, mlr3proba, distr6
* Range: [0, Inf]
* Minimize: TRUE
* Average: macro
* Parameters: eps=1e-15, se=FALSE, ERV=FALSE, na.rm=TRUE
* Properties: -
* Predict type: distr
* Return type: Score</code></pre>
</div>
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb55"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb55-1"><a href="#cb55-1" aria-hidden="true" tabindex="-1"></a>p<span class="sc">$</span><span class="fu">score</span>(rcll)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>surv.rcll 
 23.46684 </code></pre>
</div>
</div>
<div class="cell">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb57"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb57-1"><a href="#cb57-1" aria-hidden="true" tabindex="-1"></a>ibrier <span class="ot">=</span> <span class="fu">msr</span>(<span class="st">"surv.brier"</span>, <span class="at">proper =</span> <span class="cn">TRUE</span>)</span>
<span id="cb57-2"><a href="#cb57-2" aria-hidden="true" tabindex="-1"></a>ibrier</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>&lt;MeasureSurvGraf:surv.graf&gt;
* Packages: mlr3, mlr3proba
* Range: [0, Inf]
* Minimize: TRUE
* Average: macro
* Parameters: integrated=TRUE, method=2, se=FALSE, proper=TRUE,
  eps=0.001, ERV=FALSE
* Properties: -
* Predict type: distr
* Return type: Score</code></pre>
</div>
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb59"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb59-1"><a href="#cb59-1" aria-hidden="true" tabindex="-1"></a>p<span class="sc">$</span><span class="fu">score</span>(ibrier, <span class="at">task =</span> task, <span class="at">train_set =</span> part<span class="sc">$</span>train)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>surv.graf 
0.1591112 </code></pre>
</div>
</div>
</section>
</section>
<section id="using-and-tuning-ml-survival-models-on-high-dimensional-data" class="level2 unnumbered">
<h2 class="unnumbered anchored" data-anchor-id="using-and-tuning-ml-survival-models-on-high-dimensional-data">Using and tuning ML survival models on high-dimensional data</h2>
<div class="callout callout-style-default callout-tip callout-titled" title="Teaching Aims">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Teaching Aims
</div>
</div>
<div class="callout-body-container callout-body">
<ul>
<li>Create your own survival tasks from external data</li>
<li>Evaluate performance using <code>resample()</code></li>
<li>Tune hyperparameters with <code>auto_tuner()</code></li>
</ul>
</div>
</div>
<p>So far we have used the Cox regression model, but there are many more machine learning methods available via <code>mlr3extralearners</code> (<a href="https://mlr-org.com/learners.html">learner list</a>)! We will take a look at the following:</p>
<ul>
<li>Cox elastic net via <a href="https://glmnet.stanford.edu/articles/Coxnet.html"><code>glmnet</code></a> <span class="citation" data-cites="Friedman2010">(<a href="#ref-Friedman2010" role="doc-biblioref">Friedman, Hastie, and Tibshirani 2010</a>)</span>
<ul>
<li>We will use <code>lrn("surv.cv_glmnet")</code>, wich internally tunes for <code>lambda</code> using cross-validation</li>
</ul></li>
<li>Likelihood-based boosting via <a href="https://github.com/binderh/CoxBoost"><code>CoxBoost</code></a> <span class="citation" data-cites="Binder2008">(<a href="#ref-Binder2008" role="doc-biblioref">Binder and Schumacher 2008</a>)</span>
<ul>
<li>We later use <code>lrn("surv.cv_coxboost", penalty = "optimCoxBoostPenalty")</code>, which also uses internal cross-validation to tune its parameters</li>
</ul></li>
<li>Random Forests via <a href="https://imbs-hl.github.io/ranger/"><code>ranger</code></a> <span class="citation" data-cites="Ishwaran2008">(<a href="#ref-Ishwaran2008" role="doc-biblioref">Ishwaran et al. 2008</a>)</span></li>
<li>Oblique Random Forests via <a href="https://docs.ropensci.org/aorsf/"><code>aorsf</code></a> <span class="citation" data-cites="Jaeger2023">(<a href="#ref-Jaeger2023" role="doc-biblioref">Jaeger et al. 2023</a>)</span></li>
</ul>
<p>These learners then cover the range from penalized regression to tree ensembles and boosting.</p>
<p>Let’s take these learners for a spin on a subset of TCGA breast cancer data with gene expression and clinical features. We first need to create a <code>TaskSurv</code> object from the data, which we can do by reading in the data and then using <code>as_task_surv()</code>. We also add the <code>status</code> column to the stratum, which is necessary for the resampling to ensure a similar proportion of events in the resampling folds with the complete dataset.</p>
<div class="cell">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb61"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb61-1"><a href="#cb61-1" aria-hidden="true" tabindex="-1"></a>tcga <span class="ot">=</span> <span class="fu">readRDS</span>(<span class="st">"data/tcga.rds"</span>)</span>
<span id="cb61-2"><a href="#cb61-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-3"><a href="#cb61-3" aria-hidden="true" tabindex="-1"></a>task_tcga <span class="ot">=</span> mlr3proba<span class="sc">::</span><span class="fu">as_task_surv</span>(</span>
<span id="cb61-4"><a href="#cb61-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">x =</span> tcga, </span>
<span id="cb61-5"><a href="#cb61-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">time =</span> <span class="st">"time"</span>, <span class="at">event =</span> <span class="st">"status"</span>, <span class="at">id =</span> <span class="st">"BRCA-TCGA"</span></span>
<span id="cb61-6"><a href="#cb61-6" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb61-7"><a href="#cb61-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-8"><a href="#cb61-8" aria-hidden="true" tabindex="-1"></a><span class="co"># Set stratum for resampling</span></span>
<span id="cb61-9"><a href="#cb61-9" aria-hidden="true" tabindex="-1"></a>task_tcga<span class="sc">$</span><span class="fu">set_col_roles</span>(<span class="st">"status"</span>, <span class="at">add_to =</span> <span class="st">"stratum"</span>)</span>
<span id="cb61-10"><a href="#cb61-10" aria-hidden="true" tabindex="-1"></a>task_tcga</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>&lt;TaskSurv:BRCA-TCGA&gt; (1047 x 54)
* Target: time, status
* Properties: strata
* Features (52):
  - dbl (52): ACTR3B, ANLN, BAG1, BCL2, BIRC5, BLVRA, CCNB1, CCNE1,
    CDC20, CDC6, CDH3, CENPF, CEP55, CXXC5, EGFR, ERBB2, ESR1, EXO1,
    FGFR4, FOXA1, FOXC1, GPR160, GRB7, KIF2C, KRT14, KRT17, KRT5, MAPT,
    MDM2, MELK, MIA, MKI67, MLPH, MMP11, MYBL2, MYC, NAT1, NDC80, NUF2,
    ORC6, PGR, PHGDH, PTTG1, RRM2, SFRP1, SLC39A6, TMEM45B, TYMS,
    UBE2C, UBE2T, age, ethnicity
* Strata: status</code></pre>
</div>
</div>
<p>We can instantiate our learners as we’ve seen before — we’re sticking to mostly vanilla settings for now.<br>
We can let <code>glmnet</code> determine the optimal value for <code>lambda</code> with it’s internal cross-validation method Similarly, <code>CoxBoost</code> could tune itself, but we’ll stick with a simple version to save some time on compute! For the forests, we use 100 trees each for speed and otherwise accept the defaults.</p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>To speed things up a little, we let learners use 4 parallel threads (<code>num.threads</code> in <code>ranger</code> and <code>n_thread</code> in <code>aorsf</code>), which you may want to change depending on your available resources!</p>
</div>
</div>
<div class="cell">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb63"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb63-1"><a href="#cb63-1" aria-hidden="true" tabindex="-1"></a>lrn_glmnet <span class="ot">=</span> <span class="fu">lrn</span>(<span class="st">"surv.cv_glmnet"</span>, <span class="at">alpha =</span> <span class="fl">0.5</span>, <span class="at">s =</span> <span class="st">"lambda.min"</span>)</span>
<span id="cb63-2"><a href="#cb63-2" aria-hidden="true" tabindex="-1"></a>lrn_coxboost <span class="ot">=</span> <span class="fu">lrn</span>(<span class="st">"surv.coxboost"</span>, <span class="at">penalty =</span> <span class="dv">100</span>)</span>
<span id="cb63-3"><a href="#cb63-3" aria-hidden="true" tabindex="-1"></a>lrn_ranger <span class="ot">=</span> <span class="fu">lrn</span>(<span class="st">"surv.ranger"</span>, <span class="at">num.trees =</span> <span class="dv">100</span>, <span class="at">num.threads =</span> <span class="dv">4</span>)</span>
<span id="cb63-4"><a href="#cb63-4" aria-hidden="true" tabindex="-1"></a>lrn_aorsf <span class="ot">=</span> <span class="fu">lrn</span>(<span class="st">"surv.aorsf"</span>, <span class="at">n_tree =</span> <span class="dv">100</span>, <span class="at">n_thread =</span> <span class="dv">4</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>We can now use <code>resample()</code> to evaluate the performance of each of these learners on the task. To do this, we decide on two measures: Harrell’s C and the integrated brier score, and we also instantiate a resampling to use for comparison, such that we ensure all learners see the same data.</p>
<div class="cell">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb64"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb64-1"><a href="#cb64-1" aria-hidden="true" tabindex="-1"></a>measures <span class="ot">=</span> <span class="fu">list</span>(<span class="fu">msr</span>(<span class="st">"surv.cindex"</span>, <span class="at">id =</span> <span class="st">"cindex"</span>), <span class="fu">msr</span>(<span class="st">"surv.brier"</span>, <span class="at">id =</span> <span class="st">"ibs"</span>))</span>
<span id="cb64-2"><a href="#cb64-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-3"><a href="#cb64-3" aria-hidden="true" tabindex="-1"></a>resampling <span class="ot">=</span> <span class="fu">rsmp</span>(<span class="st">"cv"</span>, <span class="at">folds =</span> <span class="dv">3</span>)</span>
<span id="cb64-4"><a href="#cb64-4" aria-hidden="true" tabindex="-1"></a>resampling<span class="sc">$</span><span class="fu">instantiate</span>(task_tcga)</span>
<span id="cb64-5"><a href="#cb64-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-6"><a href="#cb64-6" aria-hidden="true" tabindex="-1"></a>rr_glmnet <span class="ot">=</span> <span class="fu">resample</span>(</span>
<span id="cb64-7"><a href="#cb64-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">task =</span> task_tcga,</span>
<span id="cb64-8"><a href="#cb64-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">learner =</span> lrn_glmnet,</span>
<span id="cb64-9"><a href="#cb64-9" aria-hidden="true" tabindex="-1"></a>  <span class="at">resampling =</span> resampling</span>
<span id="cb64-10"><a href="#cb64-10" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>INFO  [00:10:58.085] [mlr3] Applying learner 'surv.cv_glmnet' on task 'BRCA-TCGA' (iter 1/3)
INFO  [00:10:59.532] [mlr3] Applying learner 'surv.cv_glmnet' on task 'BRCA-TCGA' (iter 2/3)
INFO  [00:11:00.649] [mlr3] Applying learner 'surv.cv_glmnet' on task 'BRCA-TCGA' (iter 3/3)</code></pre>
</div>
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb66"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb66-1"><a href="#cb66-1" aria-hidden="true" tabindex="-1"></a>rr_glmnet<span class="sc">$</span><span class="fu">score</span>(measures)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>     task_id     learner_id resampling_id iteration    cindex       ibs
      &lt;char&gt;         &lt;char&gt;        &lt;char&gt;     &lt;int&gt;     &lt;num&gt;     &lt;num&gt;
1: BRCA-TCGA surv.cv_glmnet            cv         1 0.5336199 0.1428582
2: BRCA-TCGA surv.cv_glmnet            cv         2 0.7011343 0.2703538
3: BRCA-TCGA surv.cv_glmnet            cv         3 0.7201999 0.1859438
Hidden columns: task, learner, resampling, prediction</code></pre>
</div>
</div>
<p>Feel free to play with the parameters of <code>glmnet</code> a bit more — for example, does changing <code>alpha</code> help?</p>
<p>We can repeat the same procedure for the other learners:</p>
<div class="cell">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb68"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb68-1"><a href="#cb68-1" aria-hidden="true" tabindex="-1"></a>rr_coxboost <span class="ot">=</span> <span class="fu">resample</span>(</span>
<span id="cb68-2"><a href="#cb68-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">task =</span> task_tcga,</span>
<span id="cb68-3"><a href="#cb68-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">learner =</span> lrn_coxboost,</span>
<span id="cb68-4"><a href="#cb68-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">resampling =</span> resampling</span>
<span id="cb68-5"><a href="#cb68-5" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>INFO  [00:11:02.267] [mlr3] Applying learner 'surv.coxboost' on task 'BRCA-TCGA' (iter 1/3)
INFO  [00:11:02.915] [mlr3] Applying learner 'surv.coxboost' on task 'BRCA-TCGA' (iter 2/3)
INFO  [00:11:03.630] [mlr3] Applying learner 'surv.coxboost' on task 'BRCA-TCGA' (iter 3/3)</code></pre>
</div>
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb70"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb70-1"><a href="#cb70-1" aria-hidden="true" tabindex="-1"></a>rr_ranger <span class="ot">=</span> <span class="fu">resample</span>(</span>
<span id="cb70-2"><a href="#cb70-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">task =</span> task_tcga,</span>
<span id="cb70-3"><a href="#cb70-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">learner =</span> lrn_ranger,</span>
<span id="cb70-4"><a href="#cb70-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">resampling =</span> resampling</span>
<span id="cb70-5"><a href="#cb70-5" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>INFO  [00:11:04.357] [mlr3] Applying learner 'surv.ranger' on task 'BRCA-TCGA' (iter 1/3)
INFO  [00:11:05.167] [mlr3] Applying learner 'surv.ranger' on task 'BRCA-TCGA' (iter 2/3)
INFO  [00:11:05.962] [mlr3] Applying learner 'surv.ranger' on task 'BRCA-TCGA' (iter 3/3)</code></pre>
</div>
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb72"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb72-1"><a href="#cb72-1" aria-hidden="true" tabindex="-1"></a>rr_aorsf <span class="ot">=</span> <span class="fu">resample</span>(</span>
<span id="cb72-2"><a href="#cb72-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">task =</span> task_tcga,</span>
<span id="cb72-3"><a href="#cb72-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">learner =</span> lrn_aorsf,</span>
<span id="cb72-4"><a href="#cb72-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">resampling =</span> resampling</span>
<span id="cb72-5"><a href="#cb72-5" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>INFO  [00:11:06.876] [mlr3] Applying learner 'surv.aorsf' on task 'BRCA-TCGA' (iter 1/3)
INFO  [00:11:07.004] [mlr3] Applying learner 'surv.aorsf' on task 'BRCA-TCGA' (iter 2/3)
INFO  [00:11:07.062] [mlr3] Applying learner 'surv.aorsf' on task 'BRCA-TCGA' (iter 3/3)</code></pre>
</div>
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb74"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb74-1"><a href="#cb74-1" aria-hidden="true" tabindex="-1"></a>rr_coxboost<span class="sc">$</span><span class="fu">score</span>(measures)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>     task_id    learner_id resampling_id iteration    cindex       ibs
      &lt;char&gt;        &lt;char&gt;        &lt;char&gt;     &lt;int&gt;     &lt;num&gt;     &lt;num&gt;
1: BRCA-TCGA surv.coxboost            cv         1 0.5342340 0.1572601
2: BRCA-TCGA surv.coxboost            cv         2 0.6904511 0.2924077
3: BRCA-TCGA surv.coxboost            cv         3 0.6965025 0.1961894
Hidden columns: task, learner, resampling, prediction</code></pre>
</div>
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb76"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb76-1"><a href="#cb76-1" aria-hidden="true" tabindex="-1"></a>rr_ranger<span class="sc">$</span><span class="fu">score</span>(measures)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>     task_id  learner_id resampling_id iteration    cindex       ibs
      &lt;char&gt;      &lt;char&gt;        &lt;char&gt;     &lt;int&gt;     &lt;num&gt;     &lt;num&gt;
1: BRCA-TCGA surv.ranger            cv         1 0.5965613 0.1784851
2: BRCA-TCGA surv.ranger            cv         2 0.6519388 0.2638147
3: BRCA-TCGA surv.ranger            cv         3 0.6072805 0.2128354
Hidden columns: task, learner, resampling, prediction</code></pre>
</div>
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb78"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb78-1"><a href="#cb78-1" aria-hidden="true" tabindex="-1"></a>rr_aorsf<span class="sc">$</span><span class="fu">score</span>(measures)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>     task_id learner_id resampling_id iteration    cindex       ibs
      &lt;char&gt;     &lt;char&gt;        &lt;char&gt;     &lt;int&gt;     &lt;num&gt;     &lt;num&gt;
1: BRCA-TCGA surv.aorsf            cv         1 0.5784464 0.1480976
2: BRCA-TCGA surv.aorsf            cv         2 0.6875495 0.2612929
3: BRCA-TCGA surv.aorsf            cv         3 0.6519629 0.2144734
Hidden columns: task, learner, resampling, prediction</code></pre>
</div>
</div>
<p>Now we have a comparison of the performance of the different learners on the task. We can again aggregate these results to get a summary of the performance of each learner across all resamplings:</p>
<div class="cell">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb80"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb80-1"><a href="#cb80-1" aria-hidden="true" tabindex="-1"></a>rr_glmnet<span class="sc">$</span><span class="fu">aggregate</span>(measures)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>   cindex       ibs 
0.6516513 0.1997186 </code></pre>
</div>
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb82"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb82-1"><a href="#cb82-1" aria-hidden="true" tabindex="-1"></a>rr_coxboost<span class="sc">$</span><span class="fu">aggregate</span>(measures)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>   cindex       ibs 
0.6403959 0.2152857 </code></pre>
</div>
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb84"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb84-1"><a href="#cb84-1" aria-hidden="true" tabindex="-1"></a>rr_ranger<span class="sc">$</span><span class="fu">aggregate</span>(measures)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>   cindex       ibs 
0.6185935 0.2183784 </code></pre>
</div>
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb86"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb86-1"><a href="#cb86-1" aria-hidden="true" tabindex="-1"></a>rr_aorsf<span class="sc">$</span><span class="fu">aggregate</span>(measures)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>   cindex       ibs 
0.6393196 0.2079546 </code></pre>
</div>
</div>
<p>Of course in practice we want to tune these learners for optimal performance. Tuning can be quite a complex topic, but <code>mlr3</code> makes it relatively simple with the <code>auto_tuner</code> approach. Without going into too much detail about the theory, for tuning we need:</p>
<ul>
<li>A learner to tune, with information on which parameters to tune in which range (the search space)</li>
<li>A strategy (<code>tuner</code>), such as random search, grid search, or more advanced options, which defines how we search for new parameter values to try</li>
<li>A resampling strategy to evaluate performance during tuning</li>
<li>A tuning measure top optimize for</li>
<li>A stopping criterion, e.g.&nbsp;stopping after 100 evaluations</li>
</ul>
<div class="cell">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb88"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb88-1"><a href="#cb88-1" aria-hidden="true" tabindex="-1"></a>at_glmnet <span class="ot">=</span> <span class="fu">auto_tuner</span>(</span>
<span id="cb88-2"><a href="#cb88-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">learner =</span> <span class="fu">lrn</span>(<span class="st">"surv.cv_glmnet"</span>, <span class="at">alpha =</span> <span class="fu">to_tune</span>(<span class="dv">0</span>, <span class="dv">1</span>), <span class="at">s =</span> <span class="st">"lambda.min"</span>),</span>
<span id="cb88-3"><a href="#cb88-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">tuner =</span> <span class="fu">tnr</span>(<span class="st">"grid_search"</span>),</span>
<span id="cb88-4"><a href="#cb88-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">resampling =</span> <span class="fu">rsmp</span>(<span class="st">"cv"</span>, <span class="at">folds =</span> <span class="dv">3</span>),</span>
<span id="cb88-5"><a href="#cb88-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">measure =</span> <span class="fu">msr</span>(<span class="st">"surv.cindex"</span>),</span>
<span id="cb88-6"><a href="#cb88-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">term_evals =</span> <span class="dv">100</span></span>
<span id="cb88-7"><a href="#cb88-7" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>We can then try this out on the full dataset like so:</p>
<div class="cell">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb89"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb89-1"><a href="#cb89-1" aria-hidden="true" tabindex="-1"></a>at_glmnet<span class="sc">$</span><span class="fu">train</span>(task_tcga)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>INFO  [00:11:08.717] [bbotk] Starting to optimize 1 parameter(s) with '&lt;OptimizerBatchGridSearch&gt;' and '&lt;TerminatorEvals&gt; [n_evals=100, k=0]'
INFO  [00:11:08.723] [bbotk] Evaluating 1 configuration(s)
INFO  [00:11:08.728] [mlr3] Running benchmark with 3 resampling iterations
INFO  [00:11:08.732] [mlr3] Applying learner 'surv.cv_glmnet' on task 'BRCA-TCGA' (iter 1/3)
INFO  [00:11:09.912] [mlr3] Applying learner 'surv.cv_glmnet' on task 'BRCA-TCGA' (iter 2/3)
INFO  [00:11:11.305] [mlr3] Applying learner 'surv.cv_glmnet' on task 'BRCA-TCGA' (iter 3/3)
INFO  [00:11:12.364] [mlr3] Finished benchmark
INFO  [00:11:12.382] [bbotk] Result of batch 1:
INFO  [00:11:12.384] [bbotk]      alpha surv.cindex warnings errors runtime_learners
INFO  [00:11:12.384] [bbotk]  0.7777778   0.6262311        0      0            3.615
INFO  [00:11:12.384] [bbotk]                                 uhash
INFO  [00:11:12.384] [bbotk]  1c78a0de-e406-4308-9e01-35cad209a33b
INFO  [00:11:12.385] [bbotk] Evaluating 1 configuration(s)
INFO  [00:11:12.389] [mlr3] Running benchmark with 3 resampling iterations
INFO  [00:11:12.393] [mlr3] Applying learner 'surv.cv_glmnet' on task 'BRCA-TCGA' (iter 1/3)
INFO  [00:11:13.606] [mlr3] Applying learner 'surv.cv_glmnet' on task 'BRCA-TCGA' (iter 2/3)
INFO  [00:11:14.951] [mlr3] Applying learner 'surv.cv_glmnet' on task 'BRCA-TCGA' (iter 3/3)
INFO  [00:11:16.000] [mlr3] Finished benchmark
INFO  [00:11:16.018] [bbotk] Result of batch 2:
INFO  [00:11:16.019] [bbotk]  alpha surv.cindex warnings errors runtime_learners
INFO  [00:11:16.019] [bbotk]      1   0.6304686        0      0            3.592
INFO  [00:11:16.019] [bbotk]                                 uhash
INFO  [00:11:16.019] [bbotk]  bdbf7001-6698-41d2-a284-0453799cf4ac
INFO  [00:11:16.021] [bbotk] Evaluating 1 configuration(s)
INFO  [00:11:16.025] [mlr3] Running benchmark with 3 resampling iterations
INFO  [00:11:16.028] [mlr3] Applying learner 'surv.cv_glmnet' on task 'BRCA-TCGA' (iter 1/3)
INFO  [00:11:17.252] [mlr3] Applying learner 'surv.cv_glmnet' on task 'BRCA-TCGA' (iter 2/3)
INFO  [00:11:18.674] [mlr3] Applying learner 'surv.cv_glmnet' on task 'BRCA-TCGA' (iter 3/3)
INFO  [00:11:19.756] [mlr3] Finished benchmark
INFO  [00:11:19.779] [bbotk] Result of batch 3:
INFO  [00:11:19.780] [bbotk]      alpha surv.cindex warnings errors runtime_learners
INFO  [00:11:19.780] [bbotk]  0.6666667   0.6285418        0      0             3.71
INFO  [00:11:19.780] [bbotk]                                 uhash
INFO  [00:11:19.780] [bbotk]  c07db742-bdf4-4ad9-a054-4d2dfe21580c
INFO  [00:11:19.782] [bbotk] Evaluating 1 configuration(s)
INFO  [00:11:19.786] [mlr3] Running benchmark with 3 resampling iterations
INFO  [00:11:19.790] [mlr3] Applying learner 'surv.cv_glmnet' on task 'BRCA-TCGA' (iter 1/3)
INFO  [00:11:20.994] [mlr3] Applying learner 'surv.cv_glmnet' on task 'BRCA-TCGA' (iter 2/3)
INFO  [00:11:22.453] [mlr3] Applying learner 'surv.cv_glmnet' on task 'BRCA-TCGA' (iter 3/3)
INFO  [00:11:23.576] [mlr3] Finished benchmark
INFO  [00:11:23.597] [bbotk] Result of batch 4:
INFO  [00:11:23.599] [bbotk]      alpha surv.cindex warnings errors runtime_learners
INFO  [00:11:23.599] [bbotk]  0.3333333   0.6312686        0      0            3.768
INFO  [00:11:23.599] [bbotk]                                 uhash
INFO  [00:11:23.599] [bbotk]  7431e37c-d7b9-4020-b775-1e08d3bb2a18
INFO  [00:11:23.600] [bbotk] Evaluating 1 configuration(s)
INFO  [00:11:23.605] [mlr3] Running benchmark with 3 resampling iterations
INFO  [00:11:23.608] [mlr3] Applying learner 'surv.cv_glmnet' on task 'BRCA-TCGA' (iter 1/3)
INFO  [00:11:24.886] [mlr3] Applying learner 'surv.cv_glmnet' on task 'BRCA-TCGA' (iter 2/3)
INFO  [00:11:26.569] [mlr3] Applying learner 'surv.cv_glmnet' on task 'BRCA-TCGA' (iter 3/3)
INFO  [00:11:27.712] [mlr3] Finished benchmark
INFO  [00:11:27.730] [bbotk] Result of batch 5:
INFO  [00:11:27.731] [bbotk]      alpha surv.cindex warnings errors runtime_learners
INFO  [00:11:27.731] [bbotk]  0.2222222   0.6265975        0      0            4.085
INFO  [00:11:27.731] [bbotk]                                 uhash
INFO  [00:11:27.731] [bbotk]  dc2c95dd-fb53-4db5-8de4-f00e0815fa2a
INFO  [00:11:27.732] [bbotk] Evaluating 1 configuration(s)
INFO  [00:11:27.737] [mlr3] Running benchmark with 3 resampling iterations
INFO  [00:11:27.740] [mlr3] Applying learner 'surv.cv_glmnet' on task 'BRCA-TCGA' (iter 1/3)
INFO  [00:11:28.882] [mlr3] Applying learner 'surv.cv_glmnet' on task 'BRCA-TCGA' (iter 2/3)
INFO  [00:11:30.256] [mlr3] Applying learner 'surv.cv_glmnet' on task 'BRCA-TCGA' (iter 3/3)
INFO  [00:11:31.345] [mlr3] Finished benchmark
INFO  [00:11:31.367] [bbotk] Result of batch 6:
INFO  [00:11:31.369] [bbotk]      alpha surv.cindex warnings errors runtime_learners
INFO  [00:11:31.369] [bbotk]  0.5555556   0.6288861        0      0            3.589
INFO  [00:11:31.369] [bbotk]                                 uhash
INFO  [00:11:31.369] [bbotk]  f9c6d586-59e1-49de-a947-85a617bc2adb
INFO  [00:11:31.370] [bbotk] Evaluating 1 configuration(s)
INFO  [00:11:31.374] [mlr3] Running benchmark with 3 resampling iterations
INFO  [00:11:31.378] [mlr3] Applying learner 'surv.cv_glmnet' on task 'BRCA-TCGA' (iter 1/3)
INFO  [00:11:32.556] [mlr3] Applying learner 'surv.cv_glmnet' on task 'BRCA-TCGA' (iter 2/3)
INFO  [00:11:33.945] [mlr3] Applying learner 'surv.cv_glmnet' on task 'BRCA-TCGA' (iter 3/3)
INFO  [00:11:35.022] [mlr3] Finished benchmark
INFO  [00:11:35.041] [bbotk] Result of batch 7:
INFO  [00:11:35.043] [bbotk]      alpha surv.cindex warnings errors runtime_learners
INFO  [00:11:35.043] [bbotk]  0.8888889   0.6289243        0      0            3.627
INFO  [00:11:35.043] [bbotk]                                 uhash
INFO  [00:11:35.043] [bbotk]  a155997f-ba60-4d13-90be-ecc98fdd7d3d
INFO  [00:11:35.045] [bbotk] Evaluating 1 configuration(s)
INFO  [00:11:35.050] [mlr3] Running benchmark with 3 resampling iterations
INFO  [00:11:35.053] [mlr3] Applying learner 'surv.cv_glmnet' on task 'BRCA-TCGA' (iter 1/3)
INFO  [00:11:36.476] [mlr3] Applying learner 'surv.cv_glmnet' on task 'BRCA-TCGA' (iter 2/3)
INFO  [00:11:38.053] [mlr3] Applying learner 'surv.cv_glmnet' on task 'BRCA-TCGA' (iter 3/3)
INFO  [00:11:39.322] [mlr3] Finished benchmark
INFO  [00:11:39.339] [bbotk] Result of batch 8:
INFO  [00:11:39.340] [bbotk]      alpha surv.cindex warnings errors runtime_learners
INFO  [00:11:39.340] [bbotk]  0.1111111   0.6408908        0      0            4.251
INFO  [00:11:39.340] [bbotk]                                 uhash
INFO  [00:11:39.340] [bbotk]  9e72b12b-844f-4889-9995-954a91b3a3e1
INFO  [00:11:39.342] [bbotk] Evaluating 1 configuration(s)
INFO  [00:11:39.346] [mlr3] Running benchmark with 3 resampling iterations
INFO  [00:11:39.349] [mlr3] Applying learner 'surv.cv_glmnet' on task 'BRCA-TCGA' (iter 1/3)
INFO  [00:11:40.650] [mlr3] Applying learner 'surv.cv_glmnet' on task 'BRCA-TCGA' (iter 2/3)
INFO  [00:11:42.075] [mlr3] Applying learner 'surv.cv_glmnet' on task 'BRCA-TCGA' (iter 3/3)
INFO  [00:11:43.184] [mlr3] Finished benchmark
INFO  [00:11:43.207] [bbotk] Result of batch 9:
INFO  [00:11:43.208] [bbotk]      alpha surv.cindex warnings errors runtime_learners
INFO  [00:11:43.208] [bbotk]  0.4444444   0.6291064        0      0            3.816
INFO  [00:11:43.208] [bbotk]                                 uhash
INFO  [00:11:43.208] [bbotk]  32da2be8-683b-4d1d-a5fa-ed585b0ea220
INFO  [00:11:43.209] [bbotk] Evaluating 1 configuration(s)
INFO  [00:11:43.213] [mlr3] Running benchmark with 3 resampling iterations
INFO  [00:11:43.217] [mlr3] Applying learner 'surv.cv_glmnet' on task 'BRCA-TCGA' (iter 1/3)
INFO  [00:11:44.173] [mlr3] Applying learner 'surv.cv_glmnet' on task 'BRCA-TCGA' (iter 2/3)
INFO  [00:11:45.161] [mlr3] Applying learner 'surv.cv_glmnet' on task 'BRCA-TCGA' (iter 3/3)
INFO  [00:11:46.138] [mlr3] Finished benchmark
INFO  [00:11:46.155] [bbotk] Result of batch 10:
INFO  [00:11:46.157] [bbotk]  alpha surv.cindex warnings errors runtime_learners
INFO  [00:11:46.157] [bbotk]      0   0.6445429        0      0            2.904
INFO  [00:11:46.157] [bbotk]                                 uhash
INFO  [00:11:46.157] [bbotk]  a10b55d0-ca39-443e-9970-71737f6e5041
INFO  [00:11:46.164] [bbotk] Finished optimizing after 10 evaluation(s)
INFO  [00:11:46.164] [bbotk] Result:
INFO  [00:11:46.165] [bbotk]  alpha learner_param_vals  x_domain surv.cindex
INFO  [00:11:46.165] [bbotk]  &lt;num&gt;             &lt;list&gt;    &lt;list&gt;       &lt;num&gt;
INFO  [00:11:46.165] [bbotk]      0          &lt;list[2]&gt; &lt;list[1]&gt;   0.6445429</code></pre>
</div>
</div>
<p>We can see all evaluated parameter combinations in the <code>$tuning_instance</code> and the best result in <code>$tuning_result</code></p>
<div class="cell">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb91"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb91-1"><a href="#cb91-1" aria-hidden="true" tabindex="-1"></a>at_glmnet<span class="sc">$</span>tuning_instance</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>&lt;TuningInstanceBatchSingleCrit&gt;
* State:  Optimized
* Objective: &lt;ObjectiveTuningBatch:surv.cv_glmnet_on_BRCA-TCGA&gt;
* Search Space:
       id    class lower upper nlevels
   &lt;char&gt;   &lt;char&gt; &lt;num&gt; &lt;num&gt;   &lt;num&gt;
1:  alpha ParamDbl     0     1     Inf
* Terminator: &lt;TerminatorEvals&gt;
* Result:
   alpha surv.cindex
   &lt;num&gt;       &lt;num&gt;
1:     0   0.6445429
* Archive:
        alpha surv.cindex
        &lt;num&gt;       &lt;num&gt;
 1: 0.7777778   0.6262311
 2: 1.0000000   0.6304686
 3: 0.6666667   0.6285418
 4: 0.3333333   0.6312686
 5: 0.2222222   0.6265975
 6: 0.5555556   0.6288861
 7: 0.8888889   0.6289243
 8: 0.1111111   0.6408908
 9: 0.4444444   0.6291064
10: 0.0000000   0.6445429</code></pre>
</div>
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb93"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb93-1"><a href="#cb93-1" aria-hidden="true" tabindex="-1"></a>at_glmnet<span class="sc">$</span>tuning_result</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>   alpha learner_param_vals  x_domain surv.cindex
   &lt;num&gt;             &lt;list&gt;    &lt;list&gt;       &lt;num&gt;
1:     0          &lt;list[2]&gt; &lt;list[1]&gt;   0.6445429</code></pre>
</div>
</div>
<p>We get a better result than 0.5 now, but note <em>we used all of the data</em> now, so while this would be the approach we could use to find a model for new data, right now we want to compare our learners fairly! That means: <em>Nested resampling</em>, where we use resampling for tuning, and for evaluation in two layers.</p>
<p>In this step, we simplify the <code>resample()</code> steps using <code>benchmark()</code>, and we’ll also tune some of the learners.</p>
<p>We start by first defining our learners with tuning, using <code>cv_coxboost</code> to tune itself without the <code>auto_tuner</code> and tuning <code>glmnet</code> similar to before, but we’ll set small budgets to keep the runtime of this example code managable:</p>
<div class="cell">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb95"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb95-1"><a href="#cb95-1" aria-hidden="true" tabindex="-1"></a>at_glmnet <span class="ot">=</span> <span class="fu">auto_tuner</span>(</span>
<span id="cb95-2"><a href="#cb95-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">learner =</span> <span class="fu">lrn</span>(<span class="st">"surv.cv_glmnet"</span>, <span class="at">alpha =</span> <span class="fu">to_tune</span>(<span class="dv">0</span>, <span class="dv">1</span>), <span class="at">s =</span> <span class="st">"lambda.min"</span>),</span>
<span id="cb95-3"><a href="#cb95-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">tuner =</span> <span class="fu">tnr</span>(<span class="st">"grid_search"</span>),</span>
<span id="cb95-4"><a href="#cb95-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">resampling =</span> <span class="fu">rsmp</span>(<span class="st">"cv"</span>, <span class="at">folds =</span> <span class="dv">3</span>),</span>
<span id="cb95-5"><a href="#cb95-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">measure =</span> <span class="fu">msr</span>(<span class="st">"surv.cindex"</span>),</span>
<span id="cb95-6"><a href="#cb95-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">term_evals =</span> <span class="dv">25</span></span>
<span id="cb95-7"><a href="#cb95-7" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb95-8"><a href="#cb95-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb95-9"><a href="#cb95-9" aria-hidden="true" tabindex="-1"></a><span class="co"># For CoxBoost's optimCoxBoostPenalty, iter.max is analogous to term_evals </span></span>
<span id="cb95-10"><a href="#cb95-10" aria-hidden="true" tabindex="-1"></a>lrn_cvcoxboost <span class="ot">=</span> <span class="fu">lrn</span>(<span class="st">"surv.cv_coxboost"</span>, <span class="at">penalty =</span> <span class="st">"optimCoxBoostPenalty"</span>, <span class="at">iter.max =</span> <span class="dv">10</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>The we create a <em>benchmark design</em> of one or more tasks and at least two learners like so:</p>
<div class="cell">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb96"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb96-1"><a href="#cb96-1" aria-hidden="true" tabindex="-1"></a>design <span class="ot">=</span> <span class="fu">benchmark_grid</span>(</span>
<span id="cb96-2"><a href="#cb96-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">tasks =</span> task_tcga,</span>
<span id="cb96-3"><a href="#cb96-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">learners =</span> <span class="fu">list</span>(at_glmnet, lrn_ranger, lrn_aorsf, lrn_cvcoxboost),</span>
<span id="cb96-4"><a href="#cb96-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">resamplings =</span> resampling</span>
<span id="cb96-5"><a href="#cb96-5" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb96-6"><a href="#cb96-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb96-7"><a href="#cb96-7" aria-hidden="true" tabindex="-1"></a>design</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>        task              learner resampling
      &lt;char&gt;               &lt;char&gt;     &lt;char&gt;
1: BRCA-TCGA surv.cv_glmnet.tuned         cv
2: BRCA-TCGA          surv.ranger         cv
3: BRCA-TCGA           surv.aorsf         cv
4: BRCA-TCGA     surv.cv_coxboost         cv</code></pre>
</div>
</div>
<p>To perform the benchmark, we use the aptly named <code>benchmark()</code> function, which will perform the necessary resampling iterations and store the results for us — this will take a moment!</p>
<p>When we <code>$score()</code> or <code>$aggregate()</code> the benchmark result, we should get the same exact scores as before for the untuned learners because we used the instantiated resampling from earlier, meaning each learner again saw the same data — <code>glmnet</code> and <code>coxboost</code> however should do better because now we spent time tuning them!</p>
<div class="cell">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb98"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb98-1"><a href="#cb98-1" aria-hidden="true" tabindex="-1"></a><span class="co"># bmr$score(measures)</span></span>
<span id="cb98-2"><a href="#cb98-2" aria-hidden="true" tabindex="-1"></a>bmr<span class="sc">$</span><span class="fu">aggregate</span>(<span class="fu">msr</span>(<span class="st">"surv.cindex"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>We can also visualize the results — see <code>?autoplot.BenchmarkResult</code> for more options:</p>
<div class="cell">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb99"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb99-1"><a href="#cb99-1" aria-hidden="true" tabindex="-1"></a><span class="fu">autoplot</span>(bmr, <span class="at">type =</span> <span class="st">"boxplot"</span>, <span class="at">measure =</span> <span class="fu">msr</span>(<span class="st">"surv.cindex"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>From our quick tests, which learner now seems to have done the best? Given that we used these learners more or less off the shelf without tuning, we should not put too much weight on these results, but it’s a good starting point for further exploration!</p>
<p>Tuning is a complex topic we barely scratched the surface of, but you can learn more about it in the <a href="https://mlr3book.mlr-org.com/chapters/chapter4/hyperparameter_optimization.html">mlr3book chapter</a>!</p>
</section>
<section id="benchmarking-with-multiple-datasets" class="level2 unnumbered">
<h2 class="unnumbered anchored" data-anchor-id="benchmarking-with-multiple-datasets">Benchmarking with multiple datasets</h2>
<div class="callout callout-style-default callout-tip callout-titled" title="Teaching Aims">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Teaching Aims
</div>
</div>
<div class="callout-body-container callout-body">
<ul>
<li>Perform a small-scale benchmark</li>
<li>Aggregate and visualize the results</li>
<li>Perform a statistical analysis of the results</li>
</ul>
</div>
</div>
<p>A proper benchmark can take a lot of time and planning, but it can pay off to get a good overview of the performance of different learners on different tasks relevant to your field!</p>
<p>In this example, we’ll take a number of small datasets provided by <code>mlr3proba</code> and benchmark the learners we used before on them. These tasks are small enough to hopefully not spend too much time waiting for computations to finish, but we hope you get enough of an idea to feel confident to perform your own experiments!</p>
<p>The procedure is as follows:</p>
<ol type="1">
<li>Gather tasks as a list.</li>
<li>Gather our learners. Normally this would include deciding on tuning spaces!</li>
<li>Define a resampling strategy.</li>
<li>Decide on measures to use.</li>
</ol>
<p>For step 1, we’ll select some survival tasks from <code>mlr_tasks</code> for this benchmark:</p>
<div class="cell">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb100"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb100-1"><a href="#cb100-1" aria-hidden="true" tabindex="-1"></a>tasks <span class="ot">=</span> <span class="fu">list</span>(</span>
<span id="cb100-2"><a href="#cb100-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tsk</span>(<span class="st">"actg"</span>),</span>
<span id="cb100-3"><a href="#cb100-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tsk</span>(<span class="st">"gbcs"</span>),</span>
<span id="cb100-4"><a href="#cb100-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tsk</span>(<span class="st">"grace"</span>),</span>
<span id="cb100-5"><a href="#cb100-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tsk</span>(<span class="st">"lung"</span>),</span>
<span id="cb100-6"><a href="#cb100-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tsk</span>(<span class="st">"mgus"</span>)</span>
<span id="cb100-7"><a href="#cb100-7" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb100-8"><a href="#cb100-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb100-9"><a href="#cb100-9" aria-hidden="true" tabindex="-1"></a>tasks</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>[[1]]
&lt;TaskSurv:actg&gt; (1151 x 13): ACTG 320
* Target: time, status
* Properties: -
* Features (11):
  - dbl (4): age, cd4, priorzdv, sexF
  - fct (4): ivdrug, karnof, raceth, txgrp
  - int (3): hemophil, strat2, tx

[[2]]
&lt;TaskSurv:gbcs&gt; (686 x 10): German Breast Cancer
* Target: time, status
* Properties: -
* Features (8):
  - dbl (4): age, estrg_recp, prog_recp, size
  - int (4): grade, hormone, menopause, nodes

[[3]]
&lt;TaskSurv:grace&gt; (1000 x 8): GRACE 1000
* Target: time, status
* Properties: -
* Features (6):
  - dbl (4): age, los, revascdays, sysbp
  - int (2): revasc, stchange

[[4]]
&lt;TaskSurv:lung&gt; (168 x 9): Lung Cancer
* Target: time, status
* Properties: -
* Features (7):
  - int (6): age, meal.cal, pat.karno, ph.ecog, ph.karno, wt.loss
  - fct (1): sex

[[5]]
&lt;TaskSurv:mgus&gt; (176 x 9): MGUS
* Target: time, status
* Properties: -
* Features (7):
  - dbl (6): age, alb, creat, dxyr, hgb, mspike
  - fct (1): sex</code></pre>
</div>
</div>
<p>Many have categorical features (<code>fct</code>), which can be a bit tricky to handle for some learners, so we will take a shortcut and add a feature encoding <code>PipeOp</code> to the learners that need it. Pipelines and preprocessing are very useful, and <a href="https://mlr3book.mlr-org.com/chapters/chapter9/preprocessing.html#factor-encoding">the mlr3book</a> again has you covered!</p>
<p>We use the <code>po("encode")</code> pipe operator to encode the factors as dummy-encoded variables (<code>method = "treatment"</code>) for the Cox model, and use the default (one-hot encoding) for the others. The <code>%&gt;&gt;%</code> operator is used to chain <code>PipeOps</code> and learners together, and we wrap the pipeline in <code>as_learner</code> such that we can treat it as a learner just like the others.</p>
<div class="cell">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb102"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb102-1"><a href="#cb102-1" aria-hidden="true" tabindex="-1"></a>preproc <span class="ot">=</span> <span class="fu">po</span>(<span class="st">"encode"</span>, <span class="at">method =</span> <span class="st">"treatment"</span>)</span>
<span id="cb102-2"><a href="#cb102-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb102-3"><a href="#cb102-3" aria-hidden="true" tabindex="-1"></a>learners <span class="ot">=</span> <span class="fu">list</span>(</span>
<span id="cb102-4"><a href="#cb102-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">cox =</span> <span class="fu">as_learner</span>(preproc <span class="sc">%&gt;&gt;%</span> <span class="fu">lrn</span>(<span class="st">"surv.coxph"</span>,<span class="at">id =</span> <span class="st">"cph"</span>)),</span>
<span id="cb102-5"><a href="#cb102-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">glmnet =</span> <span class="fu">as_learner</span>(preproc <span class="sc">%&gt;&gt;%</span> <span class="fu">lrn</span>(<span class="st">"surv.cv_glmnet"</span>, <span class="at">alpha =</span> <span class="fl">0.5</span>)),</span>
<span id="cb102-6"><a href="#cb102-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">ranger =</span> <span class="fu">lrn</span>(<span class="st">"surv.ranger"</span>, <span class="at">num.trees =</span> <span class="dv">100</span>, <span class="at">num.threads =</span> <span class="dv">4</span>),</span>
<span id="cb102-7"><a href="#cb102-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">aorsf =</span> <span class="fu">lrn</span>(<span class="st">"surv.aorsf"</span>, <span class="at">n_tree =</span> <span class="dv">100</span>, <span class="at">n_thread =</span> <span class="dv">4</span>),</span>
<span id="cb102-8"><a href="#cb102-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">coxboost =</span> <span class="fu">as_learner</span>(preproc <span class="sc">%&gt;&gt;%</span> <span class="fu">lrn</span>(<span class="st">"surv.coxboost"</span>, <span class="at">penalty =</span> <span class="dv">100</span>))</span>
<span id="cb102-9"><a href="#cb102-9" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>A small convenience thing we can do here is to set IDs for the learners, which will make the output of further steps more readable:</p>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>$cox
[1] "cox"

$glmnet
[1] "glmnet"

$ranger
[1] "ranger"

$aorsf
[1] "aorsf"

$coxboost
[1] "coxboost"</code></pre>
</div>
</div>
<p>Moving on to the benchmark, we create a design grid as before, only now we have multiple tasks. Luckily, <code>benchmark_grid()</code> can handle this for us by instantiating the resampling for each task, so we don’t have to worry about this here!</p>
<div class="cell">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb104"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb104-1"><a href="#cb104-1" aria-hidden="true" tabindex="-1"></a>design <span class="ot">=</span> <span class="fu">benchmark_grid</span>(</span>
<span id="cb104-2"><a href="#cb104-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">tasks =</span> tasks,</span>
<span id="cb104-3"><a href="#cb104-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">learners =</span> learners,</span>
<span id="cb104-4"><a href="#cb104-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">resamplings =</span> <span class="fu">rsmp</span>(<span class="st">"cv"</span>, <span class="at">folds =</span> <span class="dv">3</span>)</span>
<span id="cb104-5"><a href="#cb104-5" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb104-6"><a href="#cb104-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb104-7"><a href="#cb104-7" aria-hidden="true" tabindex="-1"></a>design</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>      task  learner resampling
    &lt;char&gt;   &lt;char&gt;     &lt;char&gt;
 1:   actg      cox         cv
 2:   actg   glmnet         cv
 3:   actg   ranger         cv
 4:   actg    aorsf         cv
 5:   actg coxboost         cv
 6:   gbcs      cox         cv
 7:   gbcs   glmnet         cv
 8:   gbcs   ranger         cv
 9:   gbcs    aorsf         cv
10:   gbcs coxboost         cv
11:  grace      cox         cv
12:  grace   glmnet         cv
13:  grace   ranger         cv
14:  grace    aorsf         cv
15:  grace coxboost         cv
16:   lung      cox         cv
17:   lung   glmnet         cv
18:   lung   ranger         cv
19:   lung    aorsf         cv
20:   lung coxboost         cv
21:   mgus      cox         cv
22:   mgus   glmnet         cv
23:   mgus   ranger         cv
24:   mgus    aorsf         cv
25:   mgus coxboost         cv
      task  learner resampling</code></pre>
</div>
</div>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>INFO  [00:11:47.824] [mlr3] Running benchmark with 75 resampling iterations
INFO  [00:11:47.828] [mlr3] Applying learner 'cox' on task 'actg' (iter 1/3)
INFO  [00:11:47.911] [mlr3] Applying learner 'cox' on task 'actg' (iter 2/3)
INFO  [00:11:47.978] [mlr3] Applying learner 'cox' on task 'actg' (iter 3/3)
INFO  [00:11:48.239] [mlr3] Applying learner 'glmnet' on task 'actg' (iter 1/3)
INFO  [00:11:48.615] [mlr3] Applying learner 'glmnet' on task 'actg' (iter 2/3)
INFO  [00:11:49.029] [mlr3] Applying learner 'glmnet' on task 'actg' (iter 3/3)
INFO  [00:11:49.468] [mlr3] Applying learner 'ranger' on task 'actg' (iter 1/3)
INFO  [00:11:49.547] [mlr3] Applying learner 'ranger' on task 'actg' (iter 2/3)
INFO  [00:11:49.622] [mlr3] Applying learner 'ranger' on task 'actg' (iter 3/3)
INFO  [00:11:49.699] [mlr3] Applying learner 'aorsf' on task 'actg' (iter 1/3)
INFO  [00:11:49.747] [mlr3] Applying learner 'aorsf' on task 'actg' (iter 2/3)
INFO  [00:11:49.784] [mlr3] Applying learner 'aorsf' on task 'actg' (iter 3/3)
INFO  [00:11:49.821] [mlr3] Applying learner 'coxboost' on task 'actg' (iter 1/3)
INFO  [00:11:50.164] [mlr3] Applying learner 'coxboost' on task 'actg' (iter 2/3)
INFO  [00:11:50.678] [mlr3] Applying learner 'coxboost' on task 'actg' (iter 3/3)
INFO  [00:11:51.003] [mlr3] Applying learner 'cox' on task 'gbcs' (iter 1/3)
INFO  [00:11:51.047] [mlr3] Applying learner 'cox' on task 'gbcs' (iter 2/3)
INFO  [00:11:51.093] [mlr3] Applying learner 'cox' on task 'gbcs' (iter 3/3)
INFO  [00:11:51.143] [mlr3] Applying learner 'glmnet' on task 'gbcs' (iter 1/3)
INFO  [00:11:51.269] [mlr3] Applying learner 'glmnet' on task 'gbcs' (iter 2/3)
INFO  [00:11:51.394] [mlr3] Applying learner 'glmnet' on task 'gbcs' (iter 3/3)
INFO  [00:11:51.529] [mlr3] Applying learner 'ranger' on task 'gbcs' (iter 1/3)
INFO  [00:11:51.604] [mlr3] Applying learner 'ranger' on task 'gbcs' (iter 2/3)
INFO  [00:11:51.687] [mlr3] Applying learner 'ranger' on task 'gbcs' (iter 3/3)
INFO  [00:11:51.756] [mlr3] Applying learner 'aorsf' on task 'gbcs' (iter 1/3)
INFO  [00:11:51.787] [mlr3] Applying learner 'aorsf' on task 'gbcs' (iter 2/3)
INFO  [00:11:51.818] [mlr3] Applying learner 'aorsf' on task 'gbcs' (iter 3/3)
INFO  [00:11:51.847] [mlr3] Applying learner 'coxboost' on task 'gbcs' (iter 1/3)
INFO  [00:11:52.305] [mlr3] Applying learner 'coxboost' on task 'gbcs' (iter 2/3)
INFO  [00:11:52.561] [mlr3] Applying learner 'coxboost' on task 'gbcs' (iter 3/3)
INFO  [00:11:52.777] [mlr3] Applying learner 'cox' on task 'grace' (iter 1/3)
INFO  [00:11:52.819] [mlr3] Applying learner 'cox' on task 'grace' (iter 2/3)
INFO  [00:11:52.869] [mlr3] Applying learner 'cox' on task 'grace' (iter 3/3)
INFO  [00:11:52.909] [mlr3] Applying learner 'glmnet' on task 'grace' (iter 1/3)
INFO  [00:11:53.090] [mlr3] Applying learner 'glmnet' on task 'grace' (iter 2/3)
INFO  [00:11:53.266] [mlr3] Applying learner 'glmnet' on task 'grace' (iter 3/3)
INFO  [00:11:53.450] [mlr3] Applying learner 'ranger' on task 'grace' (iter 1/3)
INFO  [00:11:53.540] [mlr3] Applying learner 'ranger' on task 'grace' (iter 2/3)
INFO  [00:11:53.628] [mlr3] Applying learner 'ranger' on task 'grace' (iter 3/3)
INFO  [00:11:53.723] [mlr3] Applying learner 'aorsf' on task 'grace' (iter 1/3)
INFO  [00:11:53.762] [mlr3] Applying learner 'aorsf' on task 'grace' (iter 2/3)
INFO  [00:11:53.804] [mlr3] Applying learner 'aorsf' on task 'grace' (iter 3/3)
INFO  [00:11:53.844] [mlr3] Applying learner 'coxboost' on task 'grace' (iter 1/3)
INFO  [00:11:54.443] [mlr3] Applying learner 'coxboost' on task 'grace' (iter 2/3)
INFO  [00:11:55.058] [mlr3] Applying learner 'coxboost' on task 'grace' (iter 3/3)
INFO  [00:11:55.470] [mlr3] Applying learner 'cox' on task 'lung' (iter 1/3)
INFO  [00:11:55.523] [mlr3] Applying learner 'cox' on task 'lung' (iter 2/3)
INFO  [00:11:55.575] [mlr3] Applying learner 'cox' on task 'lung' (iter 3/3)
INFO  [00:11:55.630] [mlr3] Applying learner 'glmnet' on task 'lung' (iter 1/3)
INFO  [00:11:55.724] [mlr3] Applying learner 'glmnet' on task 'lung' (iter 2/3)
INFO  [00:11:55.818] [mlr3] Applying learner 'glmnet' on task 'lung' (iter 3/3)
INFO  [00:11:55.904] [mlr3] Applying learner 'ranger' on task 'lung' (iter 1/3)
INFO  [00:11:55.929] [mlr3] Applying learner 'ranger' on task 'lung' (iter 2/3)
INFO  [00:11:55.953] [mlr3] Applying learner 'ranger' on task 'lung' (iter 3/3)
INFO  [00:11:55.978] [mlr3] Applying learner 'aorsf' on task 'lung' (iter 1/3)
INFO  [00:11:55.999] [mlr3] Applying learner 'aorsf' on task 'lung' (iter 2/3)
INFO  [00:11:56.022] [mlr3] Applying learner 'aorsf' on task 'lung' (iter 3/3)
INFO  [00:11:56.043] [mlr3] Applying learner 'coxboost' on task 'lung' (iter 1/3)
INFO  [00:11:56.185] [mlr3] Applying learner 'coxboost' on task 'lung' (iter 2/3)
INFO  [00:11:56.317] [mlr3] Applying learner 'coxboost' on task 'lung' (iter 3/3)
INFO  [00:11:56.447] [mlr3] Applying learner 'cox' on task 'mgus' (iter 1/3)
INFO  [00:11:56.500] [mlr3] Applying learner 'cox' on task 'mgus' (iter 2/3)
INFO  [00:11:56.554] [mlr3] Applying learner 'cox' on task 'mgus' (iter 3/3)
INFO  [00:11:56.609] [mlr3] Applying learner 'glmnet' on task 'mgus' (iter 1/3)
INFO  [00:11:56.710] [mlr3] Applying learner 'glmnet' on task 'mgus' (iter 2/3)
INFO  [00:11:56.794] [mlr3] Applying learner 'glmnet' on task 'mgus' (iter 3/3)
INFO  [00:11:56.883] [mlr3] Applying learner 'ranger' on task 'mgus' (iter 1/3)
INFO  [00:11:56.912] [mlr3] Applying learner 'ranger' on task 'mgus' (iter 2/3)
INFO  [00:11:56.940] [mlr3] Applying learner 'ranger' on task 'mgus' (iter 3/3)
INFO  [00:11:56.969] [mlr3] Applying learner 'aorsf' on task 'mgus' (iter 1/3)
INFO  [00:11:57.017] [mlr3] Applying learner 'aorsf' on task 'mgus' (iter 2/3)
INFO  [00:11:57.042] [mlr3] Applying learner 'aorsf' on task 'mgus' (iter 3/3)
INFO  [00:11:57.066] [mlr3] Applying learner 'coxboost' on task 'mgus' (iter 1/3)
INFO  [00:11:57.237] [mlr3] Applying learner 'coxboost' on task 'mgus' (iter 2/3)
INFO  [00:11:57.417] [mlr3] Applying learner 'coxboost' on task 'mgus' (iter 3/3)</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in coxph.fit(X, Y, istrat, offset, init, control, weights = weights, : Loglik converged before variable  7,16,17,18 ; coefficient may be infinite. 
This happened PipeOp cph's $train()</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in coxph.fit(X, Y, istrat, offset, init, control, weights = weights, : Loglik converged before variable  7,8,17,18 ; coefficient may be infinite. 
This happened PipeOp cph's $train()</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>INFO  [00:11:57.599] [mlr3] Finished benchmark</code></pre>
</div>
</div>
<p>We pick the IBS again and aggregate the results:</p>
<div class="cell">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb110"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb110-1"><a href="#cb110-1" aria-hidden="true" tabindex="-1"></a>measure <span class="ot">=</span> <span class="fu">msr</span>(<span class="st">"surv.brier"</span>, <span class="at">id =</span> <span class="st">"ibs"</span>)</span>
<span id="cb110-2"><a href="#cb110-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb110-3"><a href="#cb110-3" aria-hidden="true" tabindex="-1"></a>bmr<span class="sc">$</span><span class="fu">aggregate</span>(measure)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>       nr task_id learner_id resampling_id iters        ibs
    &lt;int&gt;  &lt;char&gt;     &lt;char&gt;        &lt;char&gt; &lt;int&gt;      &lt;num&gt;
 1:     1    actg        cox            cv     3 0.05727819
 2:     2    actg     glmnet            cv     3 0.05940958
 3:     3    actg     ranger            cv     3 0.05774745
 4:     4    actg      aorsf            cv     3 0.05704656
 5:     5    actg   coxboost            cv     3 0.05682344
 6:     6    gbcs        cox            cv     3 0.12943443
 7:     7    gbcs     glmnet            cv     3 0.13680298
 8:     8    gbcs     ranger            cv     3 0.13765679
 9:     9    gbcs      aorsf            cv     3 0.12744115
10:    10    gbcs   coxboost            cv     3 0.12943960
11:    11   grace        cox            cv     3 0.09723106
12:    12   grace     glmnet            cv     3 0.10502925
13:    13   grace     ranger            cv     3 0.10822685
14:    14   grace      aorsf            cv     3 0.09167183
15:    15   grace   coxboost            cv     3 0.09723110
 [ reached getOption("max.print") -- omitted 11 rows ]
Hidden columns: resample_result</code></pre>
</div>
</div>
<section id="statistical-analysis" class="level3 unnumbered">
<h3 class="unnumbered anchored" data-anchor-id="statistical-analysis">Statistical analysis</h3>
<p>Rather than just computing average scores, we can leverage <code>mlr3benchmark</code> for additional analysis steps, including a statistical analysis of the results. The starting point is to convert the benchmark result (<code>bmr</code>) to an aggregated benchmark result (<code>bma</code>), which is a more convenient format for further analysis:</p>
<div class="cell">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb112"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb112-1"><a href="#cb112-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(mlr3benchmark)</span>
<span id="cb112-2"><a href="#cb112-2" aria-hidden="true" tabindex="-1"></a>bma <span class="ot">=</span> <span class="fu">as_benchmark_aggr</span>(bmr, <span class="at">meas =</span> measure)</span>
<span id="cb112-3"><a href="#cb112-3" aria-hidden="true" tabindex="-1"></a>bma</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>&lt;BenchmarkAggr&gt; of 25 rows with 5 tasks, 5 learners and 1 measure
    task_id learner_id        ibs
     &lt;fctr&gt;     &lt;fctr&gt;      &lt;num&gt;
 1:    actg        cox 0.05727819
 2:    actg     glmnet 0.05940958
 3:    actg     ranger 0.05774745
 4:    actg      aorsf 0.05704656
 5:    actg   coxboost 0.05682344
 6:    gbcs        cox 0.12943443
 7:    gbcs     glmnet 0.13680298
 8:    gbcs     ranger 0.13765679
 9:    gbcs      aorsf 0.12744115
10:    gbcs   coxboost 0.12943960
11:   grace        cox 0.09723106
12:   grace     glmnet 0.10502925
13:   grace     ranger 0.10822685
14:   grace      aorsf 0.09167183
15:   grace   coxboost 0.09723110
16:    lung        cox 0.15333368
17:    lung     glmnet 0.14658464
18:    lung     ranger 0.16722756
19:    lung      aorsf 0.14909360
20:    lung   coxboost 0.15303153
21:    mgus        cox 0.11897127
22:    mgus     glmnet 0.12739140
23:    mgus     ranger 0.13504452
24:    mgus      aorsf 0.12021372
25:    mgus   coxboost 0.11897096
    task_id learner_id        ibs</code></pre>
</div>
</div>
<p>This brings with it a few more <code>autoplot</code> methods, see <code>?autoplot.BenchmarkAggr</code>.</p>
<div class="cell">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb114"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb114-1"><a href="#cb114-1" aria-hidden="true" tabindex="-1"></a><span class="fu">autoplot</span>(bma, <span class="at">type =</span> <span class="st">"box"</span>, <span class="at">meas =</span> <span class="st">"ibs"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="index_files/figure-html/bma-box-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>For the statistical analysis, we can use a simple rank-based analysis following <span class="citation" data-cites="Demsar2006">Demšar (<a href="#ref-Demsar2006" role="doc-biblioref">2006</a>)</span> with a global Friedman test to see if there are significant differences between the learners:</p>
<div class="cell">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb115"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb115-1"><a href="#cb115-1" aria-hidden="true" tabindex="-1"></a>bma<span class="sc">$</span><span class="fu">friedman_test</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>
    Friedman rank sum test

data:  ibs and learner_id and task_id
Friedman chi-squared = 11.68, df = 4, p-value = 0.0199</code></pre>
</div>
</div>
<p>The corresponding post-hoc test for all pairwise comparison can be performed as follows:</p>
<div class="cell">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb117"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb117-1"><a href="#cb117-1" aria-hidden="true" tabindex="-1"></a>bma<span class="sc">$</span><span class="fu">friedman_posthoc</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stderr">
<pre><code>
    Pairwise comparisons using Nemenyi-Wilcoxon-Wilcox all-pairs test for a two-way balanced complete block design</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>data: ibs and learner_id and task_id</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>         cox   glmnet ranger aorsf
glmnet   0.855 -      -      -    
ranger   0.180 0.751  -      -    
aorsf    0.931 0.373  0.023  -    
coxboost 0.995 0.628  0.070  0.995</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>
P value adjustment method: single-step</code></pre>
</div>
</div>
<p>A visual approach is the critical difference plot <span class="citation" data-cites="Demsar2006">(<a href="#ref-Demsar2006" role="doc-biblioref">Demšar 2006</a>)</span>, which shows a connecting line between the learners that <em>are not</em> statistically different from each other (as far as their average ranks are concerned).</p>
<div class="cell">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb122"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb122-1"><a href="#cb122-1" aria-hidden="true" tabindex="-1"></a><span class="fu">autoplot</span>(bma, <span class="at">type =</span> <span class="st">"cd"</span>, <span class="at">meas =</span> <span class="st">"ibs"</span>, <span class="at">ratio =</span> .<span class="dv">7</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in geom_segment(aes(x = 0, xend = max(rank) + 1, y = 0, yend = 0)): All aesthetics have length 1, but the data has 5 rows.
ℹ Please consider using `annotate()` or provide this layer with data containing
  a single row.</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="index_files/figure-html/critical-difference-ibs-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
</section>
<section id="conclusion" class="level2">
<h2 class="anchored" data-anchor-id="conclusion">Conclusion</h2>
<p>We have conducted a tiny benchmark experiment on a few survival tasks using a few learners — a good starting point for further exploration! Advanced topics we did not cover in more detail include tuning and more advanced pipelines, but we hope you got a good overview of the capabilities of <code>mlr3proba</code> and <code>mlr3</code> in general.</p>
</section>
<section id="session-info" class="level2">
<h2 class="anchored" data-anchor-id="session-info">Session Info</h2>
<div class="cell">
<details open="" class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb124"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb124-1"><a href="#cb124-1" aria-hidden="true" tabindex="-1"></a>sessioninfo<span class="sc">::</span><span class="fu">session_info</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>─ Session info ───────────────────────────────────────────────────────────────
 setting  value
 version  R version 4.4.1 (2024-06-14)
 os       macOS Sonoma 14.5
 system   aarch64, darwin20
 ui       X11
 language (EN)
 collate  en_US.UTF-8
 ctype    en_US.UTF-8
 tz       Europe/Vienna
 date     2024-07-08
 pandoc   3.2.1 @ /opt/homebrew/bin/ (via rmarkdown)

─ Packages ───────────────────────────────────────────────────────────────────
 package           * version    date (UTC) lib source
 abind               1.4-5      2016-07-21 [1] CRAN (R 4.4.0)
 aorsf               0.1.5      2024-05-30 [1] CRAN (R 4.4.0)
 backports           1.5.0      2024-05-23 [1] CRAN (R 4.4.0)
 bbotk               1.0.0      2024-06-28 [1] CRAN (R 4.4.0)
 broom               1.0.6      2024-05-17 [1] CRAN (R 4.4.0)
 BWStest             0.2.3      2023-10-10 [1] CRAN (R 4.4.0)
 cachem              1.1.0      2024-05-16 [1] CRAN (R 4.4.0)
 car                 3.1-2      2023-03-30 [1] CRAN (R 4.4.0)
 carData             3.0-5      2022-01-06 [1] CRAN (R 4.4.0)
 checkmate           2.3.1      2023-12-04 [1] CRAN (R 4.4.0)
 class               7.3-22     2023-05-03 [2] CRAN (R 4.4.1)
 cli                 3.6.3      2024-06-21 [1] CRAN (R 4.4.0)
 clue                0.3-65     2023-09-23 [1] CRAN (R 4.4.0)
 cluster             2.1.6      2023-12-01 [2] CRAN (R 4.4.1)
 codetools           0.2-20     2024-03-31 [2] CRAN (R 4.4.1)
 collapse            2.0.14     2024-05-24 [1] CRAN (R 4.4.0)
 colorspace          2.1-0      2023-01-23 [1] CRAN (R 4.4.0)
 cowplot             1.1.3      2024-01-22 [1] CRAN (R 4.4.0)
 CoxBoost            1.5        2024-05-17 [1] Github (binderh/CoxBoost@1dc47d7)
 crayon              1.5.3      2024-06-20 [1] CRAN (R 4.4.0)
 data.table          1.15.4     2024-03-30 [1] CRAN (R 4.4.0)
 DEoptimR            1.1-3      2023-10-07 [1] CRAN (R 4.4.0)
 dictionar6          0.1.3      2021-09-13 [1] CRAN (R 4.4.0)
 digest              0.6.36     2024-06-23 [1] CRAN (R 4.4.0)
 diptest             0.77-1     2024-04-10 [1] CRAN (R 4.4.0)
 distr6              1.8.4      2024-05-07 [1] Github (xoopR/distr6@95d7359)
 dplyr               1.1.4      2023-11-17 [1] CRAN (R 4.4.0)
 evaluate            0.24.0     2024-06-10 [1] CRAN (R 4.4.0)
 fansi               1.0.6      2023-12-08 [1] CRAN (R 4.4.0)
 farver              2.1.2      2024-05-13 [1] CRAN (R 4.4.0)
 fastmap             1.2.0      2024-05-15 [1] CRAN (R 4.4.0)
 flexmix             2.3-19     2023-03-16 [1] CRAN (R 4.4.0)
 foreach             1.5.2      2022-02-02 [1] CRAN (R 4.4.0)
 fpc                 2.2-12     2024-04-30 [1] CRAN (R 4.4.0)
 future              1.33.2     2024-03-26 [1] CRAN (R 4.4.0)
 future.apply        1.11.2     2024-03-28 [1] CRAN (R 4.4.0)
 generics            0.1.3      2022-07-05 [1] CRAN (R 4.4.0)
 GGally              2.2.1      2024-02-14 [1] CRAN (R 4.4.0)
 ggplot2           * 3.5.1      2024-04-23 [1] CRAN (R 4.4.0)
 ggpubr              0.6.0      2023-02-10 [1] CRAN (R 4.4.0)
 ggsignif            0.6.4      2022-10-13 [1] CRAN (R 4.4.0)
 ggstats             0.6.0      2024-04-05 [1] CRAN (R 4.4.0)
 glmnet              4.1-8      2023-08-22 [1] CRAN (R 4.4.0)
 globals             0.16.3     2024-03-08 [1] CRAN (R 4.4.0)
 glue                1.7.0      2024-01-09 [1] CRAN (R 4.4.0)
 gmp                 0.7-4      2024-01-15 [1] CRAN (R 4.4.0)
 gridExtra           2.3        2017-09-09 [1] CRAN (R 4.4.0)
 gtable              0.3.5      2024-04-22 [1] CRAN (R 4.4.0)
 htmltools           0.5.8.1    2024-04-04 [1] CRAN (R 4.4.0)
 htmlwidgets         1.6.4      2023-12-06 [1] CRAN (R 4.4.0)
 iterators           1.0.14     2022-02-05 [1] CRAN (R 4.4.0)
 jsonlite            1.8.8      2023-12-04 [1] CRAN (R 4.4.0)
 kernlab             0.9-32     2023-01-31 [1] CRAN (R 4.4.0)
 km.ci               0.5-6      2022-04-06 [1] CRAN (R 4.4.0)
 KMsurv              0.1-5      2012-12-03 [1] CRAN (R 4.4.0)
 knitr               1.47       2024-05-29 [1] CRAN (R 4.4.0)
 kSamples            1.2-10     2023-10-07 [1] CRAN (R 4.4.0)
 labeling            0.4.3      2023-08-29 [1] CRAN (R 4.4.0)
 lattice             0.22-6     2024-03-20 [2] CRAN (R 4.4.1)
 lava                1.8.0      2024-03-05 [1] CRAN (R 4.4.0)
 lgr                 0.4.4      2022-09-05 [1] CRAN (R 4.4.0)
 lifecycle           1.0.4      2023-11-07 [1] CRAN (R 4.4.0)
 listenv             0.9.1      2024-01-29 [1] CRAN (R 4.4.0)
 magrittr            2.0.3      2022-03-30 [1] CRAN (R 4.4.0)
 MASS                7.3-61     2024-06-13 [1] CRAN (R 4.4.0)
 Matrix            * 1.7-0      2024-04-26 [2] CRAN (R 4.4.1)
 mclust              6.1.1      2024-04-29 [1] CRAN (R 4.4.0)
 memoise             2.0.1      2021-11-26 [1] CRAN (R 4.4.0)
 mlr3              * 0.20.0     2024-06-28 [1] CRAN (R 4.4.0)
 mlr3benchmark     * 0.1.6      2023-05-30 [1] CRAN (R 4.4.0)
 mlr3cluster         0.1.9      2024-03-18 [1] CRAN (R 4.4.0)
 mlr3data            0.7.0      2023-06-29 [1] CRAN (R 4.4.0)
 mlr3extralearners   0.8.0-9000 2024-07-02 [1] Github (mlr-org/mlr3extralearners@79220c5)
 mlr3filters         0.8.0      2024-04-10 [1] CRAN (R 4.4.0)
 mlr3fselect         1.0.0      2024-06-29 [1] CRAN (R 4.4.0)
 mlr3hyperband       0.6.0      2024-06-29 [1] CRAN (R 4.4.0)
 mlr3learners        0.7.0      2024-06-28 [1] CRAN (R 4.4.0)
 mlr3mbo             0.2.4      2024-07-06 [1] CRAN (R 4.4.0)
 mlr3measures        0.5.0      2022-08-05 [1] CRAN (R 4.4.0)
 mlr3misc            0.15.1     2024-06-24 [1] CRAN (R 4.4.0)
 mlr3pipelines       0.6.0      2024-07-01 [1] CRAN (R 4.4.0)
 mlr3proba         * 0.6.3      2024-07-03 [1] Github (mlr-org/mlr3proba@dcd4b62)
 mlr3tuning          1.0.0      2024-06-29 [1] CRAN (R 4.4.0)
 mlr3tuningspaces    0.5.1.9000 2024-07-03 [1] local
 mlr3verse         * 0.3.0      2024-06-30 [1] CRAN (R 4.4.0)
 mlr3viz             0.9.0.9000 2024-07-03 [1] Github (mlr-org/mlr3viz@aa8a86a)
 modeltools          0.2-23     2020-03-05 [1] CRAN (R 4.4.0)
 multcompView        0.1-10     2024-03-08 [1] CRAN (R 4.4.0)
 munsell             0.5.1      2024-04-01 [1] CRAN (R 4.4.0)
 mvtnorm             1.2-5      2024-05-21 [1] CRAN (R 4.4.0)
 nnet                7.3-19     2023-05-03 [2] CRAN (R 4.4.1)
 ooplah              0.2.0      2022-01-21 [1] CRAN (R 4.4.0)
 palmerpenguins      0.1.1      2022-08-15 [1] CRAN (R 4.4.0)
 paradox             1.0.0      2024-06-11 [1] CRAN (R 4.4.0)
 parallelly          1.37.1     2024-02-29 [1] CRAN (R 4.4.0)
 param6              0.2.4      2024-04-26 [1] Github (xoopR/param6@0fa3577)
 pillar              1.9.0      2023-03-22 [1] CRAN (R 4.4.0)
 pkgconfig           2.0.3      2019-09-22 [1] CRAN (R 4.4.0)
 plyr                1.8.9      2023-10-02 [1] CRAN (R 4.4.0)
 PMCMRplus           1.9.10     2023-12-10 [1] CRAN (R 4.4.0)
 prabclus            2.3-3      2023-10-24 [1] CRAN (R 4.4.0)
 pracma              2.4.4      2023-11-10 [1] CRAN (R 4.4.0)
 prodlim           * 2024.06.25 2024-06-24 [1] CRAN (R 4.4.0)
 purrr               1.0.2      2023-08-10 [1] CRAN (R 4.4.0)
 R.cache             0.16.0     2022-07-21 [1] CRAN (R 4.4.0)
 R.methodsS3         1.8.2      2022-06-13 [1] CRAN (R 4.4.0)
 R.oo                1.26.0     2024-01-24 [1] CRAN (R 4.4.0)
 R.utils             2.12.3     2023-11-18 [1] CRAN (R 4.4.0)
 R6                  2.5.1      2021-08-19 [1] CRAN (R 4.4.0)
 ranger              0.16.0     2023-11-12 [1] CRAN (R 4.4.0)
 RColorBrewer        1.1-3      2022-04-03 [1] CRAN (R 4.4.0)
 Rcpp                1.0.12     2024-01-09 [1] CRAN (R 4.4.0)
 RhpcBLASctl         0.23-42    2023-02-11 [1] CRAN (R 4.4.0)
 rlang               1.1.4      2024-06-04 [1] CRAN (R 4.4.0)
 rmarkdown           2.27       2024-05-17 [1] CRAN (R 4.4.0)
 Rmpfr               0.9-5      2024-01-21 [1] CRAN (R 4.4.0)
 robustbase          0.99-3     2024-07-01 [1] CRAN (R 4.4.0)
 rstatix             0.7.2      2023-02-01 [1] CRAN (R 4.4.0)
 scales              1.3.0      2023-11-28 [1] CRAN (R 4.4.0)
 sessioninfo         1.2.2      2021-12-06 [1] CRAN (R 4.4.0)
 set6                0.2.6      2024-04-26 [1] Github (xoopR/set6@a901255)
 shape               1.4.6.1    2024-02-23 [1] CRAN (R 4.4.0)
 spacefillr          0.3.3      2024-05-22 [1] CRAN (R 4.4.0)
 stringi             1.8.4      2024-05-06 [1] CRAN (R 4.4.0)
 stringr             1.5.1      2023-11-14 [1] CRAN (R 4.4.0)
 styler              1.10.3     2024-04-07 [1] CRAN (R 4.4.0)
 styler.mlr          0.1.0      2024-05-16 [1] Github (mlr-org/styler.mlr@8a86087)
 SuppDists           1.1-9.7    2022-01-03 [1] CRAN (R 4.4.0)
 survival          * 3.7-0      2024-06-05 [1] CRAN (R 4.4.0)
 survivalmodels      0.1.191    2024-03-19 [1] CRAN (R 4.4.0)
 survminer           0.4.9      2021-03-09 [1] CRAN (R 4.4.0)
 survMisc            0.5.6      2022-04-07 [1] CRAN (R 4.4.0)
 tibble              3.2.1      2023-03-20 [1] CRAN (R 4.4.0)
 tidyr               1.3.1      2024-01-24 [1] CRAN (R 4.4.0)
 tidyselect          1.2.1      2024-03-11 [1] CRAN (R 4.4.0)
 utf8                1.2.4      2023-10-22 [1] CRAN (R 4.4.0)
 uuid                1.2-0      2024-01-14 [1] CRAN (R 4.4.0)
 vctrs               0.6.5      2023-12-01 [1] CRAN (R 4.4.0)
 withr               3.0.0      2024-01-16 [1] CRAN (R 4.4.0)
 xfun                0.45       2024-06-16 [1] CRAN (R 4.4.0)
 xtable              1.8-4      2019-04-21 [1] CRAN (R 4.4.0)
 yaml                2.3.9      2024-07-05 [1] CRAN (R 4.4.0)
 zoo                 1.8-12     2023-04-13 [1] CRAN (R 4.4.0)

 [1] /Users/Lukas/Library/R/arm64/4.4/library
 [2] /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/library

──────────────────────────────────────────────────────────────────────────────</code></pre>
</div>
</div>



</section>

<div id="quarto-appendix" class="default"><section class="quarto-appendix-contents" role="doc-bibliography" id="quarto-bibliography"><h2 class="anchored quarto-appendix-heading">References</h2><div id="refs" class="references csl-bib-body hanging-indent" data-entry-spacing="0" role="list">
<div id="ref-Avati2020" class="csl-entry" role="listitem">
Avati, Anand, Tony Duan, Sharon Zhou, Kenneth Jung, Nigam H Shah, and Andrew Y Ng. 2020. <span>“<span class="nocase">Countdown Regression: Sharp and Calibrated Survival Predictions</span>.”</span> In <em>Proceedings of the 35th Uncertainty in Artificial Intelligence Conference</em>, edited by Ryan P Adams and Vibhav Gogate, 115:145–55. Proceedings of Machine Learning Research. PMLR. <a href="https://proceedings.mlr.press/v115/avati20a.html">https://proceedings.mlr.press/v115/avati20a.html</a>.
</div>
<div id="ref-Binder2008" class="csl-entry" role="listitem">
Binder, Harald, and Martin Schumacher. 2008. <span>“<span class="nocase">Allowing for mandatory covariates in boosting estimation of sparse high-dimensional survival models</span>.”</span> <em>BMC Bioinformatics</em> 9 (1): 1–10. <a href="https://doi.org/10.1186/1471-2105-9-14">https://doi.org/10.1186/1471-2105-9-14</a>.
</div>
<div id="ref-Demsar2006" class="csl-entry" role="listitem">
Demšar, Janez. 2006. <span>“<span class="nocase">Statistical Comparisons of Classifiers over Multiple Data Sets</span>.”</span> <em>Journal of Machine Learning Research</em> 7: 1–30. <a href="https://doi.org/10.5555/1248547.1248548">https://doi.org/10.5555/1248547.1248548</a>.
</div>
<div id="ref-Friedman2010" class="csl-entry" role="listitem">
Friedman, Jerome, Trevor Hastie, and Robert Tibshirani. 2010. <span>“Regularization Paths for Generalized Linear Models via Coordinate Descent.”</span> <em>Journal of Statistical Software</em> 33 (1): 1–22. <a href="https://doi.org/10.18637/jss.v033.i01">https://doi.org/10.18637/jss.v033.i01</a>.
</div>
<div id="ref-Graf1999" class="csl-entry" role="listitem">
Graf, Erika, Claudia Schmoor, Willi Sauerbrei, and Martin Schumacher. 1999. <span>“<span class="nocase">Assessment and comparison of prognostic classification schemes for survival data</span>.”</span> <em>Statistics in Medicine</em> 18: 2529–45. <a href="https://doi.org/10.1002/(SICI)1097-0258(19990915/30)18:17/18">https://doi.org/10.1002/(SICI)1097-0258(19990915/30)18:17/18</a>.
</div>
<div id="ref-Grambsch1994" class="csl-entry" role="listitem">
Grambsch, Patricia M, and Terry M. Therneau. 1994. <span>“<span class="nocase">Proportional hazards tests and diagnostics based on weighted residuals</span>.”</span> <em>Biometrika</em> 81 (3): 515–26. <a href="https://doi.org/10.1093/biomet/81.3.515">https://doi.org/10.1093/biomet/81.3.515</a>.
</div>
<div id="ref-Haider2020" class="csl-entry" role="listitem">
Haider, Humza, Bret Hoehn, Sarah Davis, and Russell Greiner. 2020. <span>“<span class="nocase">Effective Ways to Build and Evaluate Individual Survival Distributions</span>.”</span> <em>Journal of Machine Learning Research</em> 21 (85): 1–63. <a href="http://jmlr.org/papers/v21/18-772.html">http://jmlr.org/papers/v21/18-772.html</a>.
</div>
<div id="ref-Harrell1982" class="csl-entry" role="listitem">
Harrell, Frank E., Robert M. Califf, David B. Pryor, Kerry L. Lee, and Robert A. Rosati. 1982. <span>“<span class="nocase">Evaluating the Yield of Medical Tests</span>.”</span> <em>JAMA: The Journal of the American Medical Association</em> 247 (18): 2543–46. <a href="https://doi.org/10.1001/jama.1982.03320430047030">https://doi.org/10.1001/jama.1982.03320430047030</a>.
</div>
<div id="ref-Heagerty2005" class="csl-entry" role="listitem">
Heagerty, Patrick J., and Yingye Zheng. 2005. <span>“<span class="nocase">Survival Model Predictive Accuracy and ROC Curves</span>.”</span> <em>Biometrics</em> 61 (1): 92–105. <a href="https://doi.org/10.1111/J.0006-341X.2005.030814.X">https://doi.org/10.1111/J.0006-341X.2005.030814.X</a>.
</div>
<div id="ref-Ishwaran2008" class="csl-entry" role="listitem">
Ishwaran, Hemant, Udaya B. Kogalur, Eugene H. Blackstone, and Michael S. Lauer. 2008. <span>“<span class="nocase">Random survival forests</span>”</span> 2 (3): 841–60.
</div>
<div id="ref-Jaeger2023" class="csl-entry" role="listitem">
Jaeger, Byron C., Sawyer Welden, Kristin Lenoir, Jaime L. Speiser, Matthew W. Segar, Ambarish Pandey, and Nicholas M. Pajewski. 2023. <span>“<span class="nocase">Accelerated and Interpretable Oblique Random Survival Forests</span>.”</span> <em>Journal of Computational and Graphical Statistics</em>, 1–16. <a href="https://doi.org/10.1080/10618600.2023.2231048">https://doi.org/10.1080/10618600.2023.2231048</a>.
</div>
<div id="ref-mlr3" class="csl-entry" role="listitem">
Lang, Michel, Martin Binder, Jakob Richter, Patrick Schratz, Florian Pfisterer, Stefan Coors, Quay Au, Giuseppe Casalicchio, Lars Kotthoff, and Bernd Bischl. 2019. <span>“<span class="nocase">mlr3: A modern object-oriented machine learning framework in R</span>.”</span> <em>Journal of Open Source Software</em> 4 (44): 1903. <a href="https://doi.org/10.21105/JOSS.01903">https://doi.org/10.21105/JOSS.01903</a>.
</div>
<div id="ref-mlr3proba" class="csl-entry" role="listitem">
Sonabend, Raphael, Franz J. Király, Andreas Bender, Bernd Bischl, and Michel Lang. 2021. <span>“<span class="nocase">mlr3proba: an R package for machine learning in survival analysis</span>.”</span> <em>Bioinformatics</em> 37 (17): 2789–91. <a href="https://doi.org/10.1093/BIOINFORMATICS/BTAB039">https://doi.org/10.1093/BIOINFORMATICS/BTAB039</a>.
</div>
<div id="ref-distr6" class="csl-entry" role="listitem">
Sonabend, Raphael, and Franz J. Kirdly. 2021. <span>“<span class="nocase">distr6: R6 Object-Oriented Probability Distributions Interface in R</span>.”</span> <em>R Journal</em> 13 (1): 444–66. <a href="https://doi.org/10.32614/RJ-2021-055">https://doi.org/10.32614/RJ-2021-055</a>.
</div>
<div id="ref-Uno2011" class="csl-entry" role="listitem">
Uno, Hajime, Tianxi Cai, Michael J. Pencina, Ralph B. D’Agostino, and L. J. Wei. 2011. <span>“<span class="nocase">On the C-statistics for evaluating overall adequacy of risk prediction procedures with censored survival data</span>.”</span> <em>Statistics in Medicine</em> 30 (10): 1105–17. <a href="https://doi.org/10.1002/SIM.4154">https://doi.org/10.1002/SIM.4154</a>.
</div>
<div id="ref-Uno2007" class="csl-entry" role="listitem">
Uno, Hajime, Tianxi Cai, Lu Tian, and L. J. Wei. 2007. <span>“<span class="nocase">Evaluating Prediction Rules for t-Year Survivors With Censored Regression Models</span>.”</span> <em>Journal of the American Statistical Association</em> 102 (478): 527–37. <a href="https://doi.org/10.1198/016214507000000149">https://doi.org/10.1198/016214507000000149</a>.
</div>
</div></section></div></main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    // For code content inside modals, clipBoardJS needs to be initialized with a container option
    // TODO: Check when it could be a function (https://github.com/zenorocha/clipboard.js/issues/860)
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->




<footer class="footer"><div class="nav-footer"><div class="nav-footer-center"><div class="toc-actions d-sm-block d-md-none"><ul><li><a href="https://github.com/survival-org/user2024-survival-mlr3/blob/main/index.qmd" class="toc-action"><i class="bi bi-github"></i>View source</a></li></ul></div></div></div></footer></body></html>